{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4cf17e00-550a-44c8-a88a-96217e5ad1ff",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'__name__': '__main__',\n",
       " '__doc__': 'Automatically created module for IPython interactive environment',\n",
       " '__package__': None,\n",
       " '__loader__': None,\n",
       " '__spec__': None,\n",
       " '__builtin__': <module 'builtins' (built-in)>,\n",
       " '__builtins__': <module 'builtins' (built-in)>,\n",
       " '_ih': ['',\n",
       "  '# import packages\\n\\nimport tissue.main, tissue.downstream\\n\\nimport numpy as np\\nimport pandas as pd\\nimport matplotlib.pyplot as plt\\nimport scanpy as sc\\nimport anndata as ad\\nimport os',\n",
       "  '# loading\\nsc_data=sc.read_h5ad(\"/data/kjc2/projects/P330.CSA/rds/downsampled_100_25-04-15-08-34.h5ad\")\\nst_data=sc.read_h5ad(\"/data/kjc2/projects/P330.CSA/rds/250210_CRC_BJM_0050585_Region1_25-04-14-16-53.h5ad\")',\n",
       "  'tissue.main.preprocess_data(sc_data, standardize=False, normalize=False)',\n",
       "  'gene_names = np.intersect1d(st_data.var_names, sc_data.var_names)\\nst_data = st_data[:, gene_names].copy()',\n",
       "  '# hold out target gene\\ntarget_gene = \"plp1\"\\ntarget_expn = st_data[:, target_gene].X.copy()\\nst_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()',\n",
       "  '# hold out target gene\\ntarget_gene = \"PLP1\"\\ntarget_expn = st_data[:, target_gene].X.copy()\\nst_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()',\n",
       "  'print(gene_names)',\n",
       "  '# hold out target gene\\ntarget_gene = \"ZYX\"\\ntarget_expn = st_data[:, target_gene].X.copy()\\nst_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()',\n",
       "  'sc_data.shape\\nst_data.shape',\n",
       "  'sc_data.shape',\n",
       "  'print(sc_data.shape)\\nprint(st_data.shape)',\n",
       "  '# SpaGE spatial gene expression prediction\\n\\ntissue.main.predict_gene_expression (st_data, sc_data, [target_gene],\\n                                     method=\"spage\", n_folds=10, n_pv=10)',\n",
       "  'np.isnan(sc_data.X).sum()',\n",
       "  'import scipy.sparse\\nnp.isnan(adata.X.A).sum() if scipy.sparse.issparse(adata.X) else np.isnan(adata.X).sum()',\n",
       "  'import scipy.sparse\\nnp.isnan(sc_data.X.A).sum() if scipy.sparse.issparse(sc_data.X) else np.isnan(sc_data.X).sum()',\n",
       "  'sc_data.obs.isna().sum()  # 각 열별 NA 개수\\nsc_data.var.isna().sum()',\n",
       "  'st_data.obs.isna().sum()  # 각 열별 NA 개수\\nst_data.var.isna().sum()',\n",
       "  'st_data.obs[st_data.obs.isna().any(axis=1)]\\nst_data.var[st_data.var.isna().any(axis=1)]',\n",
       "  'np.isnan(st_data.layers[\"counts\"]).sum() #특정 layers의 NA\\nst_data.obsm[\\'X_pca\\'][np.isnan(st_data.obsm[\\'X_pca\\']).any(axis=1)] #.obsm, .varm 등에서도.',\n",
       "  'print(st_data)',\n",
       "  'print(sc_data)\\nprint(st_data)',\n",
       "  'st_data.keys',\n",
       "  'st_data.obs.isna().sum()',\n",
       "  'st_data.obs[st_data.obs.isna().any(axis=1)] #NA가 있는 행들 보기',\n",
       "  'st_data_original=st_datast_data = st_data[~st_data.obs.isna().any(axis=1), :]\\nst_data = st_data[:, ~st_data.var.isna().any(axis=1)]',\n",
       "  'globals()'],\n",
       " '_oh': {9: (62941, 4959),\n",
       "  10: (42649, 28476),\n",
       "  16: var_names                           0\n",
       "  ensembl                             0\n",
       "  Geneid                              0\n",
       "  GeneSymbol                          0\n",
       "  Chromosome                          0\n",
       "  Start                               0\n",
       "  End                                 0\n",
       "  Class                               0\n",
       "  Strand                              0\n",
       "  Length                              0\n",
       "  Version                             0\n",
       "  Dataset_25pct_Overlap               0\n",
       "  n_cells                             0\n",
       "  n_counts                            0\n",
       "  n_cells_by_counts                   0\n",
       "  mean_counts                         0\n",
       "  total_counts                        0\n",
       "  highly_variable                     0\n",
       "  original_ensembl_id_with_version    0\n",
       "  base_ensembl_id                     0\n",
       "  dtype: int64,\n",
       "  17: gene_ids                    0\n",
       "  feature_types               0\n",
       "  genome                      0\n",
       "  n_cells_by_counts           0\n",
       "  mean_counts                 0\n",
       "  pct_dropout_by_counts       0\n",
       "  total_counts                0\n",
       "  n_cells                     0\n",
       "  highly_variable             0\n",
       "  highly_variable_rank     2972\n",
       "  means                       0\n",
       "  variances                   0\n",
       "  variances_norm              0\n",
       "  mean                        0\n",
       "  std                         0\n",
       "  dtype: int64,\n",
       "  18:                 gene_ids    feature_types   genome  n_cells_by_counts  \\\n",
       "  aamp     ENSG00000127837  Gene Expression  Unknown               9066   \n",
       "  aar2     ENSG00000131043  Gene Expression  Unknown               2083   \n",
       "  aarsd1   ENSG00000266967  Gene Expression  Unknown               2605   \n",
       "  abat     ENSG00000183044  Gene Expression  Unknown               1219   \n",
       "  abcb6    ENSG00000115657  Gene Expression  Unknown               1536   \n",
       "  ...                  ...              ...      ...                ...   \n",
       "  zscan16  ENSG00000196812  Gene Expression  Unknown               1398   \n",
       "  zscan20  ENSG00000121903  Gene Expression  Unknown                772   \n",
       "  zswim6   ENSG00000130449  Gene Expression  Unknown               1632   \n",
       "  zup1     ENSG00000153975  Gene Expression  Unknown               1595   \n",
       "  zyg11b   ENSG00000162378  Gene Expression  Unknown               2666   \n",
       "  \n",
       "           mean_counts  pct_dropout_by_counts  total_counts  n_cells  \\\n",
       "  aamp        0.161963              86.597679       10956.0     9023   \n",
       "  aar2        0.032863              96.920689        2223.0     2069   \n",
       "  aarsd1      0.040846              96.149013        2763.0     2590   \n",
       "  abat        0.019174              98.197945        1297.0     1212   \n",
       "  abcb6       0.023756              97.729322        1607.0     1528   \n",
       "  ...              ...                    ...           ...      ...   \n",
       "  zscan16     0.021790              97.933328        1474.0     1392   \n",
       "  zscan20     0.011826              98.858748         800.0      770   \n",
       "  zswim6      0.025057              97.587405        1695.0     1623   \n",
       "  zup1        0.024599              97.642102        1664.0     1585   \n",
       "  zyg11b      0.041156              96.058837        2784.0     2656   \n",
       "  \n",
       "           highly_variable  highly_variable_rank     means  variances  \\\n",
       "  aamp               False                   NaN  0.462755   1.342914   \n",
       "  aar2               False                   NaN  0.105981   0.347680   \n",
       "  aarsd1             False                   NaN  0.129163   0.408590   \n",
       "  abat               False                   NaN  0.060924   0.197949   \n",
       "  abcb6              False                   NaN  0.076851   0.250312   \n",
       "  ...                  ...                   ...       ...        ...   \n",
       "  zscan16            False                   NaN  0.069670   0.224528   \n",
       "  zscan20            False                   NaN  0.040093   0.135733   \n",
       "  zswim6             False                   NaN  0.080299   0.256135   \n",
       "  zup1               False                   NaN  0.079940   0.259380   \n",
       "  zyg11b             False                   NaN  0.135292   0.435584   \n",
       "  \n",
       "           variances_norm      mean       std  \n",
       "  aamp           0.966223  0.462755  1.158842  \n",
       "  aar2           0.979476  0.105981  0.589644  \n",
       "  aarsd1         0.960307  0.129163  0.639210  \n",
       "  abat           0.947347  0.060924  0.444915  \n",
       "  abcb6          0.956212  0.076851  0.500312  \n",
       "  ...                 ...       ...       ...  \n",
       "  zscan16        0.943187  0.069670  0.473844  \n",
       "  zscan20        0.973417  0.040093  0.368420  \n",
       "  zswim6         0.937661  0.080299  0.506098  \n",
       "  zup1           0.953678  0.079940  0.509293  \n",
       "  zyg11b         0.980638  0.135292  0.659988  \n",
       "  \n",
       "  [2972 rows x 15 columns],\n",
       "  23: x_centroid                       0\n",
       "  y_centroid                       0\n",
       "  transcript_counts                0\n",
       "  control_probe_counts             0\n",
       "  genomic_control_counts           0\n",
       "  control_codeword_counts          0\n",
       "  unassigned_codeword_counts       0\n",
       "  deprecated_codeword_counts       0\n",
       "  total_counts                     0\n",
       "  cell_area                        0\n",
       "  nucleus_area                  1695\n",
       "  nucleus_count                    0\n",
       "  segmentation_method              0\n",
       "  n_genes_by_counts                0\n",
       "  n_genes                          0\n",
       "  n_counts                         0\n",
       "  leiden_res0.8                    0\n",
       "  dtype: int64,\n",
       "  24:             x_centroid   y_centroid  transcript_counts  control_probe_counts  \\\n",
       "  aacmaokk-1  405.386169  2441.927979                459                     0   \n",
       "  aadaljne-1  368.016602  2380.937744                 57                     0   \n",
       "  aaefagpo-1  351.166016  3623.459229                256                     0   \n",
       "  aaehngjp-1  337.040192  3655.487793                266                     0   \n",
       "  aaekadjc-1  344.637512  3629.710938                348                     0   \n",
       "  ...                ...          ...                ...                   ...   \n",
       "  djhekopl-1  545.984497  4863.115234                460                     0   \n",
       "  djhjicnp-1  551.421509  4860.001465                478                     0   \n",
       "  djhkeoke-1  555.476868  4852.318848                461                     0   \n",
       "  djigmoec-1  523.969360  4901.240723                483                     0   \n",
       "  djjchnhc-1  567.563660  4861.365723                491                     0   \n",
       "  \n",
       "              genomic_control_counts  control_codeword_counts  \\\n",
       "  aacmaokk-1                       0                        0   \n",
       "  aadaljne-1                       0                        0   \n",
       "  aaefagpo-1                       0                        0   \n",
       "  aaehngjp-1                       0                        0   \n",
       "  aaekadjc-1                       0                        0   \n",
       "  ...                            ...                      ...   \n",
       "  djhekopl-1                       0                        0   \n",
       "  djhjicnp-1                       0                        0   \n",
       "  djhkeoke-1                       0                        0   \n",
       "  djigmoec-1                       0                        0   \n",
       "  djjchnhc-1                       0                        0   \n",
       "  \n",
       "              unassigned_codeword_counts  deprecated_codeword_counts  \\\n",
       "  aacmaokk-1                           0                           1   \n",
       "  aadaljne-1                           0                           0   \n",
       "  aaefagpo-1                           0                           2   \n",
       "  aaehngjp-1                           0                           0   \n",
       "  aaekadjc-1                           0                           0   \n",
       "  ...                                ...                         ...   \n",
       "  djhekopl-1                           0                           0   \n",
       "  djhjicnp-1                           0                           0   \n",
       "  djhkeoke-1                           0                           0   \n",
       "  djigmoec-1                           0                           1   \n",
       "  djjchnhc-1                           0                           0   \n",
       "  \n",
       "              total_counts  cell_area  nucleus_area  nucleus_count  \\\n",
       "  aacmaokk-1         459.0  40.730939           NaN              0   \n",
       "  aadaljne-1          57.0  20.681563           NaN              0   \n",
       "  aaefagpo-1         256.0  26.551876           NaN              0   \n",
       "  aaehngjp-1         266.0  23.390938           NaN              0   \n",
       "  aaekadjc-1         348.0  29.577345           NaN              0   \n",
       "  ...                  ...        ...           ...            ...   \n",
       "  djhekopl-1         460.0  30.209532           NaN              0   \n",
       "  djhjicnp-1         478.0  28.403282           NaN              0   \n",
       "  djhkeoke-1         461.0  38.157033           NaN              0   \n",
       "  djigmoec-1         483.0  28.448439           NaN              0   \n",
       "  djjchnhc-1         491.0  31.338439           NaN              0   \n",
       "  \n",
       "                                            segmentation_method  \\\n",
       "  aacmaokk-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  aadaljne-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  aaefagpo-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  aaehngjp-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  aaekadjc-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  ...                                                       ...   \n",
       "  djhekopl-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  djhjicnp-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  djhkeoke-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  djigmoec-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  djjchnhc-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  \n",
       "              n_genes_by_counts  n_genes  n_counts leiden_res0.8  \n",
       "  aacmaokk-1                362      362     459.0             5  \n",
       "  aadaljne-1                 50       50      57.0             2  \n",
       "  aaefagpo-1                229      229     256.0             2  \n",
       "  aaehngjp-1                220      220     266.0             2  \n",
       "  aaekadjc-1                295      295     348.0             2  \n",
       "  ...                       ...      ...       ...           ...  \n",
       "  djhekopl-1                339      339     460.0             0  \n",
       "  djhjicnp-1                366      366     478.0             0  \n",
       "  djhkeoke-1                331      331     461.0             0  \n",
       "  djigmoec-1                344      344     483.0             0  \n",
       "  djjchnhc-1                359      359     491.0             0  \n",
       "  \n",
       "  [1695 rows x 17 columns]},\n",
       " '_dh': [PosixPath('/home/jaecheon/pyscripts')],\n",
       " 'In': ['',\n",
       "  '# import packages\\n\\nimport tissue.main, tissue.downstream\\n\\nimport numpy as np\\nimport pandas as pd\\nimport matplotlib.pyplot as plt\\nimport scanpy as sc\\nimport anndata as ad\\nimport os',\n",
       "  '# loading\\nsc_data=sc.read_h5ad(\"/data/kjc2/projects/P330.CSA/rds/downsampled_100_25-04-15-08-34.h5ad\")\\nst_data=sc.read_h5ad(\"/data/kjc2/projects/P330.CSA/rds/250210_CRC_BJM_0050585_Region1_25-04-14-16-53.h5ad\")',\n",
       "  'tissue.main.preprocess_data(sc_data, standardize=False, normalize=False)',\n",
       "  'gene_names = np.intersect1d(st_data.var_names, sc_data.var_names)\\nst_data = st_data[:, gene_names].copy()',\n",
       "  '# hold out target gene\\ntarget_gene = \"plp1\"\\ntarget_expn = st_data[:, target_gene].X.copy()\\nst_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()',\n",
       "  '# hold out target gene\\ntarget_gene = \"PLP1\"\\ntarget_expn = st_data[:, target_gene].X.copy()\\nst_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()',\n",
       "  'print(gene_names)',\n",
       "  '# hold out target gene\\ntarget_gene = \"ZYX\"\\ntarget_expn = st_data[:, target_gene].X.copy()\\nst_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()',\n",
       "  'sc_data.shape\\nst_data.shape',\n",
       "  'sc_data.shape',\n",
       "  'print(sc_data.shape)\\nprint(st_data.shape)',\n",
       "  '# SpaGE spatial gene expression prediction\\n\\ntissue.main.predict_gene_expression (st_data, sc_data, [target_gene],\\n                                     method=\"spage\", n_folds=10, n_pv=10)',\n",
       "  'np.isnan(sc_data.X).sum()',\n",
       "  'import scipy.sparse\\nnp.isnan(adata.X.A).sum() if scipy.sparse.issparse(adata.X) else np.isnan(adata.X).sum()',\n",
       "  'import scipy.sparse\\nnp.isnan(sc_data.X.A).sum() if scipy.sparse.issparse(sc_data.X) else np.isnan(sc_data.X).sum()',\n",
       "  'sc_data.obs.isna().sum()  # 각 열별 NA 개수\\nsc_data.var.isna().sum()',\n",
       "  'st_data.obs.isna().sum()  # 각 열별 NA 개수\\nst_data.var.isna().sum()',\n",
       "  'st_data.obs[st_data.obs.isna().any(axis=1)]\\nst_data.var[st_data.var.isna().any(axis=1)]',\n",
       "  'np.isnan(st_data.layers[\"counts\"]).sum() #특정 layers의 NA\\nst_data.obsm[\\'X_pca\\'][np.isnan(st_data.obsm[\\'X_pca\\']).any(axis=1)] #.obsm, .varm 등에서도.',\n",
       "  'print(st_data)',\n",
       "  'print(sc_data)\\nprint(st_data)',\n",
       "  'st_data.keys',\n",
       "  'st_data.obs.isna().sum()',\n",
       "  'st_data.obs[st_data.obs.isna().any(axis=1)] #NA가 있는 행들 보기',\n",
       "  'st_data_original=st_datast_data = st_data[~st_data.obs.isna().any(axis=1), :]\\nst_data = st_data[:, ~st_data.var.isna().any(axis=1)]',\n",
       "  'globals()'],\n",
       " 'Out': {9: (62941, 4959),\n",
       "  10: (42649, 28476),\n",
       "  16: var_names                           0\n",
       "  ensembl                             0\n",
       "  Geneid                              0\n",
       "  GeneSymbol                          0\n",
       "  Chromosome                          0\n",
       "  Start                               0\n",
       "  End                                 0\n",
       "  Class                               0\n",
       "  Strand                              0\n",
       "  Length                              0\n",
       "  Version                             0\n",
       "  Dataset_25pct_Overlap               0\n",
       "  n_cells                             0\n",
       "  n_counts                            0\n",
       "  n_cells_by_counts                   0\n",
       "  mean_counts                         0\n",
       "  total_counts                        0\n",
       "  highly_variable                     0\n",
       "  original_ensembl_id_with_version    0\n",
       "  base_ensembl_id                     0\n",
       "  dtype: int64,\n",
       "  17: gene_ids                    0\n",
       "  feature_types               0\n",
       "  genome                      0\n",
       "  n_cells_by_counts           0\n",
       "  mean_counts                 0\n",
       "  pct_dropout_by_counts       0\n",
       "  total_counts                0\n",
       "  n_cells                     0\n",
       "  highly_variable             0\n",
       "  highly_variable_rank     2972\n",
       "  means                       0\n",
       "  variances                   0\n",
       "  variances_norm              0\n",
       "  mean                        0\n",
       "  std                         0\n",
       "  dtype: int64,\n",
       "  18:                 gene_ids    feature_types   genome  n_cells_by_counts  \\\n",
       "  aamp     ENSG00000127837  Gene Expression  Unknown               9066   \n",
       "  aar2     ENSG00000131043  Gene Expression  Unknown               2083   \n",
       "  aarsd1   ENSG00000266967  Gene Expression  Unknown               2605   \n",
       "  abat     ENSG00000183044  Gene Expression  Unknown               1219   \n",
       "  abcb6    ENSG00000115657  Gene Expression  Unknown               1536   \n",
       "  ...                  ...              ...      ...                ...   \n",
       "  zscan16  ENSG00000196812  Gene Expression  Unknown               1398   \n",
       "  zscan20  ENSG00000121903  Gene Expression  Unknown                772   \n",
       "  zswim6   ENSG00000130449  Gene Expression  Unknown               1632   \n",
       "  zup1     ENSG00000153975  Gene Expression  Unknown               1595   \n",
       "  zyg11b   ENSG00000162378  Gene Expression  Unknown               2666   \n",
       "  \n",
       "           mean_counts  pct_dropout_by_counts  total_counts  n_cells  \\\n",
       "  aamp        0.161963              86.597679       10956.0     9023   \n",
       "  aar2        0.032863              96.920689        2223.0     2069   \n",
       "  aarsd1      0.040846              96.149013        2763.0     2590   \n",
       "  abat        0.019174              98.197945        1297.0     1212   \n",
       "  abcb6       0.023756              97.729322        1607.0     1528   \n",
       "  ...              ...                    ...           ...      ...   \n",
       "  zscan16     0.021790              97.933328        1474.0     1392   \n",
       "  zscan20     0.011826              98.858748         800.0      770   \n",
       "  zswim6      0.025057              97.587405        1695.0     1623   \n",
       "  zup1        0.024599              97.642102        1664.0     1585   \n",
       "  zyg11b      0.041156              96.058837        2784.0     2656   \n",
       "  \n",
       "           highly_variable  highly_variable_rank     means  variances  \\\n",
       "  aamp               False                   NaN  0.462755   1.342914   \n",
       "  aar2               False                   NaN  0.105981   0.347680   \n",
       "  aarsd1             False                   NaN  0.129163   0.408590   \n",
       "  abat               False                   NaN  0.060924   0.197949   \n",
       "  abcb6              False                   NaN  0.076851   0.250312   \n",
       "  ...                  ...                   ...       ...        ...   \n",
       "  zscan16            False                   NaN  0.069670   0.224528   \n",
       "  zscan20            False                   NaN  0.040093   0.135733   \n",
       "  zswim6             False                   NaN  0.080299   0.256135   \n",
       "  zup1               False                   NaN  0.079940   0.259380   \n",
       "  zyg11b             False                   NaN  0.135292   0.435584   \n",
       "  \n",
       "           variances_norm      mean       std  \n",
       "  aamp           0.966223  0.462755  1.158842  \n",
       "  aar2           0.979476  0.105981  0.589644  \n",
       "  aarsd1         0.960307  0.129163  0.639210  \n",
       "  abat           0.947347  0.060924  0.444915  \n",
       "  abcb6          0.956212  0.076851  0.500312  \n",
       "  ...                 ...       ...       ...  \n",
       "  zscan16        0.943187  0.069670  0.473844  \n",
       "  zscan20        0.973417  0.040093  0.368420  \n",
       "  zswim6         0.937661  0.080299  0.506098  \n",
       "  zup1           0.953678  0.079940  0.509293  \n",
       "  zyg11b         0.980638  0.135292  0.659988  \n",
       "  \n",
       "  [2972 rows x 15 columns],\n",
       "  23: x_centroid                       0\n",
       "  y_centroid                       0\n",
       "  transcript_counts                0\n",
       "  control_probe_counts             0\n",
       "  genomic_control_counts           0\n",
       "  control_codeword_counts          0\n",
       "  unassigned_codeword_counts       0\n",
       "  deprecated_codeword_counts       0\n",
       "  total_counts                     0\n",
       "  cell_area                        0\n",
       "  nucleus_area                  1695\n",
       "  nucleus_count                    0\n",
       "  segmentation_method              0\n",
       "  n_genes_by_counts                0\n",
       "  n_genes                          0\n",
       "  n_counts                         0\n",
       "  leiden_res0.8                    0\n",
       "  dtype: int64,\n",
       "  24:             x_centroid   y_centroid  transcript_counts  control_probe_counts  \\\n",
       "  aacmaokk-1  405.386169  2441.927979                459                     0   \n",
       "  aadaljne-1  368.016602  2380.937744                 57                     0   \n",
       "  aaefagpo-1  351.166016  3623.459229                256                     0   \n",
       "  aaehngjp-1  337.040192  3655.487793                266                     0   \n",
       "  aaekadjc-1  344.637512  3629.710938                348                     0   \n",
       "  ...                ...          ...                ...                   ...   \n",
       "  djhekopl-1  545.984497  4863.115234                460                     0   \n",
       "  djhjicnp-1  551.421509  4860.001465                478                     0   \n",
       "  djhkeoke-1  555.476868  4852.318848                461                     0   \n",
       "  djigmoec-1  523.969360  4901.240723                483                     0   \n",
       "  djjchnhc-1  567.563660  4861.365723                491                     0   \n",
       "  \n",
       "              genomic_control_counts  control_codeword_counts  \\\n",
       "  aacmaokk-1                       0                        0   \n",
       "  aadaljne-1                       0                        0   \n",
       "  aaefagpo-1                       0                        0   \n",
       "  aaehngjp-1                       0                        0   \n",
       "  aaekadjc-1                       0                        0   \n",
       "  ...                            ...                      ...   \n",
       "  djhekopl-1                       0                        0   \n",
       "  djhjicnp-1                       0                        0   \n",
       "  djhkeoke-1                       0                        0   \n",
       "  djigmoec-1                       0                        0   \n",
       "  djjchnhc-1                       0                        0   \n",
       "  \n",
       "              unassigned_codeword_counts  deprecated_codeword_counts  \\\n",
       "  aacmaokk-1                           0                           1   \n",
       "  aadaljne-1                           0                           0   \n",
       "  aaefagpo-1                           0                           2   \n",
       "  aaehngjp-1                           0                           0   \n",
       "  aaekadjc-1                           0                           0   \n",
       "  ...                                ...                         ...   \n",
       "  djhekopl-1                           0                           0   \n",
       "  djhjicnp-1                           0                           0   \n",
       "  djhkeoke-1                           0                           0   \n",
       "  djigmoec-1                           0                           1   \n",
       "  djjchnhc-1                           0                           0   \n",
       "  \n",
       "              total_counts  cell_area  nucleus_area  nucleus_count  \\\n",
       "  aacmaokk-1         459.0  40.730939           NaN              0   \n",
       "  aadaljne-1          57.0  20.681563           NaN              0   \n",
       "  aaefagpo-1         256.0  26.551876           NaN              0   \n",
       "  aaehngjp-1         266.0  23.390938           NaN              0   \n",
       "  aaekadjc-1         348.0  29.577345           NaN              0   \n",
       "  ...                  ...        ...           ...            ...   \n",
       "  djhekopl-1         460.0  30.209532           NaN              0   \n",
       "  djhjicnp-1         478.0  28.403282           NaN              0   \n",
       "  djhkeoke-1         461.0  38.157033           NaN              0   \n",
       "  djigmoec-1         483.0  28.448439           NaN              0   \n",
       "  djjchnhc-1         491.0  31.338439           NaN              0   \n",
       "  \n",
       "                                            segmentation_method  \\\n",
       "  aacmaokk-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  aadaljne-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  aaefagpo-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  aaehngjp-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  aaekadjc-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  ...                                                       ...   \n",
       "  djhekopl-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  djhjicnp-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  djhkeoke-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  djigmoec-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  djjchnhc-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "  \n",
       "              n_genes_by_counts  n_genes  n_counts leiden_res0.8  \n",
       "  aacmaokk-1                362      362     459.0             5  \n",
       "  aadaljne-1                 50       50      57.0             2  \n",
       "  aaefagpo-1                229      229     256.0             2  \n",
       "  aaehngjp-1                220      220     266.0             2  \n",
       "  aaekadjc-1                295      295     348.0             2  \n",
       "  ...                       ...      ...       ...           ...  \n",
       "  djhekopl-1                339      339     460.0             0  \n",
       "  djhjicnp-1                366      366     478.0             0  \n",
       "  djhkeoke-1                331      331     461.0             0  \n",
       "  djigmoec-1                344      344     483.0             0  \n",
       "  djjchnhc-1                359      359     491.0             0  \n",
       "  \n",
       "  [1695 rows x 17 columns]},\n",
       " 'get_ipython': <bound method InteractiveShell.get_ipython of <ipykernel.zmqshell.ZMQInteractiveShell object at 0x7f5513e69660>>,\n",
       " 'exit': <IPython.core.autocall.ZMQExitAutocall at 0x7f5513e6a2f0>,\n",
       " 'quit': <IPython.core.autocall.ZMQExitAutocall at 0x7f5513e6a2f0>,\n",
       " 'open': <function io.open(file, mode='r', buffering=-1, encoding=None, errors=None, newline=None, closefd=True, opener=None)>,\n",
       " '_':             x_centroid   y_centroid  transcript_counts  control_probe_counts  \\\n",
       " aacmaokk-1  405.386169  2441.927979                459                     0   \n",
       " aadaljne-1  368.016602  2380.937744                 57                     0   \n",
       " aaefagpo-1  351.166016  3623.459229                256                     0   \n",
       " aaehngjp-1  337.040192  3655.487793                266                     0   \n",
       " aaekadjc-1  344.637512  3629.710938                348                     0   \n",
       " ...                ...          ...                ...                   ...   \n",
       " djhekopl-1  545.984497  4863.115234                460                     0   \n",
       " djhjicnp-1  551.421509  4860.001465                478                     0   \n",
       " djhkeoke-1  555.476868  4852.318848                461                     0   \n",
       " djigmoec-1  523.969360  4901.240723                483                     0   \n",
       " djjchnhc-1  567.563660  4861.365723                491                     0   \n",
       " \n",
       "             genomic_control_counts  control_codeword_counts  \\\n",
       " aacmaokk-1                       0                        0   \n",
       " aadaljne-1                       0                        0   \n",
       " aaefagpo-1                       0                        0   \n",
       " aaehngjp-1                       0                        0   \n",
       " aaekadjc-1                       0                        0   \n",
       " ...                            ...                      ...   \n",
       " djhekopl-1                       0                        0   \n",
       " djhjicnp-1                       0                        0   \n",
       " djhkeoke-1                       0                        0   \n",
       " djigmoec-1                       0                        0   \n",
       " djjchnhc-1                       0                        0   \n",
       " \n",
       "             unassigned_codeword_counts  deprecated_codeword_counts  \\\n",
       " aacmaokk-1                           0                           1   \n",
       " aadaljne-1                           0                           0   \n",
       " aaefagpo-1                           0                           2   \n",
       " aaehngjp-1                           0                           0   \n",
       " aaekadjc-1                           0                           0   \n",
       " ...                                ...                         ...   \n",
       " djhekopl-1                           0                           0   \n",
       " djhjicnp-1                           0                           0   \n",
       " djhkeoke-1                           0                           0   \n",
       " djigmoec-1                           0                           1   \n",
       " djjchnhc-1                           0                           0   \n",
       " \n",
       "             total_counts  cell_area  nucleus_area  nucleus_count  \\\n",
       " aacmaokk-1         459.0  40.730939           NaN              0   \n",
       " aadaljne-1          57.0  20.681563           NaN              0   \n",
       " aaefagpo-1         256.0  26.551876           NaN              0   \n",
       " aaehngjp-1         266.0  23.390938           NaN              0   \n",
       " aaekadjc-1         348.0  29.577345           NaN              0   \n",
       " ...                  ...        ...           ...            ...   \n",
       " djhekopl-1         460.0  30.209532           NaN              0   \n",
       " djhjicnp-1         478.0  28.403282           NaN              0   \n",
       " djhkeoke-1         461.0  38.157033           NaN              0   \n",
       " djigmoec-1         483.0  28.448439           NaN              0   \n",
       " djjchnhc-1         491.0  31.338439           NaN              0   \n",
       " \n",
       "                                           segmentation_method  \\\n",
       " aacmaokk-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " aadaljne-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " aaefagpo-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " aaehngjp-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " aaekadjc-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " ...                                                       ...   \n",
       " djhekopl-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " djhjicnp-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " djhkeoke-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " djigmoec-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " djjchnhc-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " \n",
       "             n_genes_by_counts  n_genes  n_counts leiden_res0.8  \n",
       " aacmaokk-1                362      362     459.0             5  \n",
       " aadaljne-1                 50       50      57.0             2  \n",
       " aaefagpo-1                229      229     256.0             2  \n",
       " aaehngjp-1                220      220     266.0             2  \n",
       " aaekadjc-1                295      295     348.0             2  \n",
       " ...                       ...      ...       ...           ...  \n",
       " djhekopl-1                339      339     460.0             0  \n",
       " djhjicnp-1                366      366     478.0             0  \n",
       " djhkeoke-1                331      331     461.0             0  \n",
       " djigmoec-1                344      344     483.0             0  \n",
       " djjchnhc-1                359      359     491.0             0  \n",
       " \n",
       " [1695 rows x 17 columns],\n",
       " '__': x_centroid                       0\n",
       " y_centroid                       0\n",
       " transcript_counts                0\n",
       " control_probe_counts             0\n",
       " genomic_control_counts           0\n",
       " control_codeword_counts          0\n",
       " unassigned_codeword_counts       0\n",
       " deprecated_codeword_counts       0\n",
       " total_counts                     0\n",
       " cell_area                        0\n",
       " nucleus_area                  1695\n",
       " nucleus_count                    0\n",
       " segmentation_method              0\n",
       " n_genes_by_counts                0\n",
       " n_genes                          0\n",
       " n_counts                         0\n",
       " leiden_res0.8                    0\n",
       " dtype: int64,\n",
       " '___':                 gene_ids    feature_types   genome  n_cells_by_counts  \\\n",
       " aamp     ENSG00000127837  Gene Expression  Unknown               9066   \n",
       " aar2     ENSG00000131043  Gene Expression  Unknown               2083   \n",
       " aarsd1   ENSG00000266967  Gene Expression  Unknown               2605   \n",
       " abat     ENSG00000183044  Gene Expression  Unknown               1219   \n",
       " abcb6    ENSG00000115657  Gene Expression  Unknown               1536   \n",
       " ...                  ...              ...      ...                ...   \n",
       " zscan16  ENSG00000196812  Gene Expression  Unknown               1398   \n",
       " zscan20  ENSG00000121903  Gene Expression  Unknown                772   \n",
       " zswim6   ENSG00000130449  Gene Expression  Unknown               1632   \n",
       " zup1     ENSG00000153975  Gene Expression  Unknown               1595   \n",
       " zyg11b   ENSG00000162378  Gene Expression  Unknown               2666   \n",
       " \n",
       "          mean_counts  pct_dropout_by_counts  total_counts  n_cells  \\\n",
       " aamp        0.161963              86.597679       10956.0     9023   \n",
       " aar2        0.032863              96.920689        2223.0     2069   \n",
       " aarsd1      0.040846              96.149013        2763.0     2590   \n",
       " abat        0.019174              98.197945        1297.0     1212   \n",
       " abcb6       0.023756              97.729322        1607.0     1528   \n",
       " ...              ...                    ...           ...      ...   \n",
       " zscan16     0.021790              97.933328        1474.0     1392   \n",
       " zscan20     0.011826              98.858748         800.0      770   \n",
       " zswim6      0.025057              97.587405        1695.0     1623   \n",
       " zup1        0.024599              97.642102        1664.0     1585   \n",
       " zyg11b      0.041156              96.058837        2784.0     2656   \n",
       " \n",
       "          highly_variable  highly_variable_rank     means  variances  \\\n",
       " aamp               False                   NaN  0.462755   1.342914   \n",
       " aar2               False                   NaN  0.105981   0.347680   \n",
       " aarsd1             False                   NaN  0.129163   0.408590   \n",
       " abat               False                   NaN  0.060924   0.197949   \n",
       " abcb6              False                   NaN  0.076851   0.250312   \n",
       " ...                  ...                   ...       ...        ...   \n",
       " zscan16            False                   NaN  0.069670   0.224528   \n",
       " zscan20            False                   NaN  0.040093   0.135733   \n",
       " zswim6             False                   NaN  0.080299   0.256135   \n",
       " zup1               False                   NaN  0.079940   0.259380   \n",
       " zyg11b             False                   NaN  0.135292   0.435584   \n",
       " \n",
       "          variances_norm      mean       std  \n",
       " aamp           0.966223  0.462755  1.158842  \n",
       " aar2           0.979476  0.105981  0.589644  \n",
       " aarsd1         0.960307  0.129163  0.639210  \n",
       " abat           0.947347  0.060924  0.444915  \n",
       " abcb6          0.956212  0.076851  0.500312  \n",
       " ...                 ...       ...       ...  \n",
       " zscan16        0.943187  0.069670  0.473844  \n",
       " zscan20        0.973417  0.040093  0.368420  \n",
       " zswim6         0.937661  0.080299  0.506098  \n",
       " zup1           0.953678  0.079940  0.509293  \n",
       " zyg11b         0.980638  0.135292  0.659988  \n",
       " \n",
       " [2972 rows x 15 columns],\n",
       " '__session__': '/home/jaecheon/pyscripts/Untitled.ipynb',\n",
       " '_i': 'st_data_original=st_datast_data = st_data[~st_data.obs.isna().any(axis=1), :]\\nst_data = st_data[:, ~st_data.var.isna().any(axis=1)]',\n",
       " '_ii': 'st_data.obs[st_data.obs.isna().any(axis=1)] #NA가 있는 행들 보기',\n",
       " '_iii': 'st_data.obs.isna().sum()',\n",
       " '_i1': '# import packages\\n\\nimport tissue.main, tissue.downstream\\n\\nimport numpy as np\\nimport pandas as pd\\nimport matplotlib.pyplot as plt\\nimport scanpy as sc\\nimport anndata as ad\\nimport os',\n",
       " 'tissue': <module 'tissue' from '/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/__init__.py'>,\n",
       " 'np': <module 'numpy' from '/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/__init__.py'>,\n",
       " 'pd': <module 'pandas' from '/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/pandas/__init__.py'>,\n",
       " 'plt': <module 'matplotlib.pyplot' from '/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/matplotlib/pyplot.py'>,\n",
       " 'sc': <module 'scanpy' from '/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/scanpy/__init__.py'>,\n",
       " 'ad': <module 'anndata' from '/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/anndata/__init__.py'>,\n",
       " 'os': <module 'os' from '/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/os.py'>,\n",
       " '_i2': '# loading\\nsc_data=sc.read_h5ad(\"/data/kjc2/projects/P330.CSA/rds/downsampled_100_25-04-15-08-34.h5ad\")\\nst_data=sc.read_h5ad(\"/data/kjc2/projects/P330.CSA/rds/250210_CRC_BJM_0050585_Region1_25-04-14-16-53.h5ad\")',\n",
       " 'sc_data': AnnData object with n_obs × n_vars = 42649 × 28476\n",
       "     obs: 'dataset', 'medical_condition', 'cancer_type', 'sample_id', 'sample_type', 'tumor_source', 'replicate', 'sample_tissue', 'anatomic_region', 'anatomic_location', 'tumor_stage', 'tumor_stage_TNM', 'tumor_stage_TNM_T', 'tumor_stage_TNM_N', 'tumor_stage_TNM_M', 'tumor_size', 'tumor_dimensions', 'tumor_grade', 'histological_type', 'microsatellite_status', 'mismatch_repair_deficiency_status', 'MLH1_promoter_methylation_status', 'MLH1_status', 'KRAS_status', 'BRAF_status', 'APC_status', 'TP53_status', 'PIK3CA_status', 'SMAD4_status', 'NRAS_status', 'MSH6_status', 'FBXW7_status', 'NOTCH1_status', 'MSH2_status', 'PMS2_status', 'POLE_status', 'ERBB2_status', 'STK11_status', 'HER2_status', 'CTNNB1_status', 'BRAS_status', 'patient_id', 'sex', 'age', 'treatment_status_before_resection', 'treatment_drug', 'treatment_response', 'RECIST', 'platform', 'platform_fine', 'cellranger_version', 'reference_genome', 'matrix_type', 'enrichment_cell_types', 'tissue_cell_state', 'tissue_processing_lab', 'hospital_location', 'country', 'NCBI_BioProject_accession', 'SRA_sample_accession', 'GEO_sample_accession', 'ENA_sample_accession', 'synapse_sample_accession', 'study_id', 'study_doi', 'study_pmid', 'original_obs_names', 'cell_type_coarse_study', 'cell_type_middle_study', 'cell_type_study', 'n_counts', 'n_genes', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_20_genes', 'pct_counts_mito', 'S_score', 'G2M_score', 'phase', 'SOLO_doublet_prob', 'SOLO_singlet_prob', 'SOLO_doublet_status', 'cell_type_predicted', 'cell_type_coarse', 'cell_type_middle', 'cell_type_fine', 'CMS_type', 'immune_infiltration_type', 'is_primary_data', 'suspension_type', 'tissue_type', 'donor_id', 'disease', 'disease_ontology_term_id', 'assay', 'assay_ontology_term_id', 'tissue', 'tissue_ontology_term_id', 'sex_ontology_term_id', 'self_reported_ethnicity', 'self_reported_ethnicity_ontology_term_id', 'organism', 'organism_ontology_term_id', 'development_stage', 'development_stage_ontology_term_id', 'cell_type', 'cell_type_ontology_term_id'\n",
       "     var: 'var_names', 'ensembl', 'Geneid', 'GeneSymbol', 'Chromosome', 'Start', 'End', 'Class', 'Strand', 'Length', 'Version', 'Dataset_25pct_Overlap', 'n_cells', 'n_counts', 'n_cells_by_counts', 'mean_counts', 'total_counts', 'highly_variable', 'original_ensembl_id_with_version', 'base_ensembl_id'\n",
       "     obsm: 'X_scANVI', 'X_scVI', 'X_umap'\n",
       "     layers: 'counts',\n",
       " 'st_data': View of AnnData object with n_obs × n_vars = 62941 × 1987\n",
       "     obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'n_genes_by_counts', 'n_genes', 'n_counts', 'leiden_res0.8'\n",
       "     var: 'gene_ids', 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'mean', 'std'\n",
       "     uns: 'hvg', 'leiden_res0.8', 'leiden_res0.8_colors', 'log1p', 'neighbors', 'pca', 'umap'\n",
       "     obsm: 'X_pca', 'X_umap', 'spatial'\n",
       "     varm: 'PCs'\n",
       "     obsp: 'connectivities', 'distances',\n",
       " '_i3': 'tissue.main.preprocess_data(sc_data, standardize=False, normalize=False)',\n",
       " '_i4': 'gene_names = np.intersect1d(st_data.var_names, sc_data.var_names)\\nst_data = st_data[:, gene_names].copy()',\n",
       " 'gene_names': array(['A2ML1', 'AAMP', 'AAR2', ..., 'ZUP1', 'ZYG11B', 'ZYX'],\n",
       "       shape=(4960,), dtype=object),\n",
       " '_i5': '# hold out target gene\\ntarget_gene = \"plp1\"\\ntarget_expn = st_data[:, target_gene].X.copy()\\nst_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()',\n",
       " 'target_gene': 'ZYX',\n",
       " '_i6': '# hold out target gene\\ntarget_gene = \"PLP1\"\\ntarget_expn = st_data[:, target_gene].X.copy()\\nst_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()',\n",
       " '_i7': 'print(gene_names)',\n",
       " '_i8': '# hold out target gene\\ntarget_gene = \"ZYX\"\\ntarget_expn = st_data[:, target_gene].X.copy()\\nst_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()',\n",
       " 'target_expn': array([[-0.35807884],\n",
       "        [-0.35807884],\n",
       "        [-0.35807884],\n",
       "        ...,\n",
       "        [-0.35807884],\n",
       "        [ 3.886576  ],\n",
       "        [-0.35807884]], shape=(62941, 1), dtype=float32),\n",
       " '_i9': 'sc_data.shape\\nst_data.shape',\n",
       " '_9': (62941, 4959),\n",
       " '_i10': 'sc_data.shape',\n",
       " '_10': (42649, 28476),\n",
       " '_i11': 'print(sc_data.shape)\\nprint(st_data.shape)',\n",
       " '_i12': '# SpaGE spatial gene expression prediction\\n\\ntissue.main.predict_gene_expression (st_data, sc_data, [target_gene],\\n                                     method=\"spage\", n_folds=10, n_pv=10)',\n",
       " '_i13': 'np.isnan(sc_data.X).sum()',\n",
       " '_i14': 'import scipy.sparse\\nnp.isnan(adata.X.A).sum() if scipy.sparse.issparse(adata.X) else np.isnan(adata.X).sum()',\n",
       " 'scipy': <module 'scipy' from '/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/scipy/__init__.py'>,\n",
       " '_i15': 'import scipy.sparse\\nnp.isnan(sc_data.X.A).sum() if scipy.sparse.issparse(sc_data.X) else np.isnan(sc_data.X).sum()',\n",
       " '_i16': 'sc_data.obs.isna().sum()  # 각 열별 NA 개수\\nsc_data.var.isna().sum()',\n",
       " '_16': var_names                           0\n",
       " ensembl                             0\n",
       " Geneid                              0\n",
       " GeneSymbol                          0\n",
       " Chromosome                          0\n",
       " Start                               0\n",
       " End                                 0\n",
       " Class                               0\n",
       " Strand                              0\n",
       " Length                              0\n",
       " Version                             0\n",
       " Dataset_25pct_Overlap               0\n",
       " n_cells                             0\n",
       " n_counts                            0\n",
       " n_cells_by_counts                   0\n",
       " mean_counts                         0\n",
       " total_counts                        0\n",
       " highly_variable                     0\n",
       " original_ensembl_id_with_version    0\n",
       " base_ensembl_id                     0\n",
       " dtype: int64,\n",
       " '_i17': 'st_data.obs.isna().sum()  # 각 열별 NA 개수\\nst_data.var.isna().sum()',\n",
       " '_17': gene_ids                    0\n",
       " feature_types               0\n",
       " genome                      0\n",
       " n_cells_by_counts           0\n",
       " mean_counts                 0\n",
       " pct_dropout_by_counts       0\n",
       " total_counts                0\n",
       " n_cells                     0\n",
       " highly_variable             0\n",
       " highly_variable_rank     2972\n",
       " means                       0\n",
       " variances                   0\n",
       " variances_norm              0\n",
       " mean                        0\n",
       " std                         0\n",
       " dtype: int64,\n",
       " '_i18': 'st_data.obs[st_data.obs.isna().any(axis=1)]\\nst_data.var[st_data.var.isna().any(axis=1)]',\n",
       " '_18':                 gene_ids    feature_types   genome  n_cells_by_counts  \\\n",
       " aamp     ENSG00000127837  Gene Expression  Unknown               9066   \n",
       " aar2     ENSG00000131043  Gene Expression  Unknown               2083   \n",
       " aarsd1   ENSG00000266967  Gene Expression  Unknown               2605   \n",
       " abat     ENSG00000183044  Gene Expression  Unknown               1219   \n",
       " abcb6    ENSG00000115657  Gene Expression  Unknown               1536   \n",
       " ...                  ...              ...      ...                ...   \n",
       " zscan16  ENSG00000196812  Gene Expression  Unknown               1398   \n",
       " zscan20  ENSG00000121903  Gene Expression  Unknown                772   \n",
       " zswim6   ENSG00000130449  Gene Expression  Unknown               1632   \n",
       " zup1     ENSG00000153975  Gene Expression  Unknown               1595   \n",
       " zyg11b   ENSG00000162378  Gene Expression  Unknown               2666   \n",
       " \n",
       "          mean_counts  pct_dropout_by_counts  total_counts  n_cells  \\\n",
       " aamp        0.161963              86.597679       10956.0     9023   \n",
       " aar2        0.032863              96.920689        2223.0     2069   \n",
       " aarsd1      0.040846              96.149013        2763.0     2590   \n",
       " abat        0.019174              98.197945        1297.0     1212   \n",
       " abcb6       0.023756              97.729322        1607.0     1528   \n",
       " ...              ...                    ...           ...      ...   \n",
       " zscan16     0.021790              97.933328        1474.0     1392   \n",
       " zscan20     0.011826              98.858748         800.0      770   \n",
       " zswim6      0.025057              97.587405        1695.0     1623   \n",
       " zup1        0.024599              97.642102        1664.0     1585   \n",
       " zyg11b      0.041156              96.058837        2784.0     2656   \n",
       " \n",
       "          highly_variable  highly_variable_rank     means  variances  \\\n",
       " aamp               False                   NaN  0.462755   1.342914   \n",
       " aar2               False                   NaN  0.105981   0.347680   \n",
       " aarsd1             False                   NaN  0.129163   0.408590   \n",
       " abat               False                   NaN  0.060924   0.197949   \n",
       " abcb6              False                   NaN  0.076851   0.250312   \n",
       " ...                  ...                   ...       ...        ...   \n",
       " zscan16            False                   NaN  0.069670   0.224528   \n",
       " zscan20            False                   NaN  0.040093   0.135733   \n",
       " zswim6             False                   NaN  0.080299   0.256135   \n",
       " zup1               False                   NaN  0.079940   0.259380   \n",
       " zyg11b             False                   NaN  0.135292   0.435584   \n",
       " \n",
       "          variances_norm      mean       std  \n",
       " aamp           0.966223  0.462755  1.158842  \n",
       " aar2           0.979476  0.105981  0.589644  \n",
       " aarsd1         0.960307  0.129163  0.639210  \n",
       " abat           0.947347  0.060924  0.444915  \n",
       " abcb6          0.956212  0.076851  0.500312  \n",
       " ...                 ...       ...       ...  \n",
       " zscan16        0.943187  0.069670  0.473844  \n",
       " zscan20        0.973417  0.040093  0.368420  \n",
       " zswim6         0.937661  0.080299  0.506098  \n",
       " zup1           0.953678  0.079940  0.509293  \n",
       " zyg11b         0.980638  0.135292  0.659988  \n",
       " \n",
       " [2972 rows x 15 columns],\n",
       " '_i19': 'np.isnan(st_data.layers[\"counts\"]).sum() #특정 layers의 NA\\nst_data.obsm[\\'X_pca\\'][np.isnan(st_data.obsm[\\'X_pca\\']).any(axis=1)] #.obsm, .varm 등에서도.',\n",
       " '_i20': 'print(st_data)',\n",
       " '_i21': 'print(sc_data)\\nprint(st_data)',\n",
       " '_i22': 'st_data.keys',\n",
       " '_i23': 'st_data.obs.isna().sum()',\n",
       " '_23': x_centroid                       0\n",
       " y_centroid                       0\n",
       " transcript_counts                0\n",
       " control_probe_counts             0\n",
       " genomic_control_counts           0\n",
       " control_codeword_counts          0\n",
       " unassigned_codeword_counts       0\n",
       " deprecated_codeword_counts       0\n",
       " total_counts                     0\n",
       " cell_area                        0\n",
       " nucleus_area                  1695\n",
       " nucleus_count                    0\n",
       " segmentation_method              0\n",
       " n_genes_by_counts                0\n",
       " n_genes                          0\n",
       " n_counts                         0\n",
       " leiden_res0.8                    0\n",
       " dtype: int64,\n",
       " '_i24': 'st_data.obs[st_data.obs.isna().any(axis=1)] #NA가 있는 행들 보기',\n",
       " '_24':             x_centroid   y_centroid  transcript_counts  control_probe_counts  \\\n",
       " aacmaokk-1  405.386169  2441.927979                459                     0   \n",
       " aadaljne-1  368.016602  2380.937744                 57                     0   \n",
       " aaefagpo-1  351.166016  3623.459229                256                     0   \n",
       " aaehngjp-1  337.040192  3655.487793                266                     0   \n",
       " aaekadjc-1  344.637512  3629.710938                348                     0   \n",
       " ...                ...          ...                ...                   ...   \n",
       " djhekopl-1  545.984497  4863.115234                460                     0   \n",
       " djhjicnp-1  551.421509  4860.001465                478                     0   \n",
       " djhkeoke-1  555.476868  4852.318848                461                     0   \n",
       " djigmoec-1  523.969360  4901.240723                483                     0   \n",
       " djjchnhc-1  567.563660  4861.365723                491                     0   \n",
       " \n",
       "             genomic_control_counts  control_codeword_counts  \\\n",
       " aacmaokk-1                       0                        0   \n",
       " aadaljne-1                       0                        0   \n",
       " aaefagpo-1                       0                        0   \n",
       " aaehngjp-1                       0                        0   \n",
       " aaekadjc-1                       0                        0   \n",
       " ...                            ...                      ...   \n",
       " djhekopl-1                       0                        0   \n",
       " djhjicnp-1                       0                        0   \n",
       " djhkeoke-1                       0                        0   \n",
       " djigmoec-1                       0                        0   \n",
       " djjchnhc-1                       0                        0   \n",
       " \n",
       "             unassigned_codeword_counts  deprecated_codeword_counts  \\\n",
       " aacmaokk-1                           0                           1   \n",
       " aadaljne-1                           0                           0   \n",
       " aaefagpo-1                           0                           2   \n",
       " aaehngjp-1                           0                           0   \n",
       " aaekadjc-1                           0                           0   \n",
       " ...                                ...                         ...   \n",
       " djhekopl-1                           0                           0   \n",
       " djhjicnp-1                           0                           0   \n",
       " djhkeoke-1                           0                           0   \n",
       " djigmoec-1                           0                           1   \n",
       " djjchnhc-1                           0                           0   \n",
       " \n",
       "             total_counts  cell_area  nucleus_area  nucleus_count  \\\n",
       " aacmaokk-1         459.0  40.730939           NaN              0   \n",
       " aadaljne-1          57.0  20.681563           NaN              0   \n",
       " aaefagpo-1         256.0  26.551876           NaN              0   \n",
       " aaehngjp-1         266.0  23.390938           NaN              0   \n",
       " aaekadjc-1         348.0  29.577345           NaN              0   \n",
       " ...                  ...        ...           ...            ...   \n",
       " djhekopl-1         460.0  30.209532           NaN              0   \n",
       " djhjicnp-1         478.0  28.403282           NaN              0   \n",
       " djhkeoke-1         461.0  38.157033           NaN              0   \n",
       " djigmoec-1         483.0  28.448439           NaN              0   \n",
       " djjchnhc-1         491.0  31.338439           NaN              0   \n",
       " \n",
       "                                           segmentation_method  \\\n",
       " aacmaokk-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " aadaljne-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " aaefagpo-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " aaehngjp-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " aaekadjc-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " ...                                                       ...   \n",
       " djhekopl-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " djhjicnp-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " djhkeoke-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " djigmoec-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " djjchnhc-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       " \n",
       "             n_genes_by_counts  n_genes  n_counts leiden_res0.8  \n",
       " aacmaokk-1                362      362     459.0             5  \n",
       " aadaljne-1                 50       50      57.0             2  \n",
       " aaefagpo-1                229      229     256.0             2  \n",
       " aaehngjp-1                220      220     266.0             2  \n",
       " aaekadjc-1                295      295     348.0             2  \n",
       " ...                       ...      ...       ...           ...  \n",
       " djhekopl-1                339      339     460.0             0  \n",
       " djhjicnp-1                366      366     478.0             0  \n",
       " djhkeoke-1                331      331     461.0             0  \n",
       " djigmoec-1                344      344     483.0             0  \n",
       " djjchnhc-1                359      359     491.0             0  \n",
       " \n",
       " [1695 rows x 17 columns],\n",
       " '_i25': 'st_data_original=st_datast_data = st_data[~st_data.obs.isna().any(axis=1), :]\\nst_data = st_data[:, ~st_data.var.isna().any(axis=1)]',\n",
       " 'st_data_original': View of AnnData object with n_obs × n_vars = 61246 × 4959\n",
       "     obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'n_genes_by_counts', 'n_genes', 'n_counts', 'leiden_res0.8'\n",
       "     var: 'gene_ids', 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'mean', 'std'\n",
       "     uns: 'hvg', 'leiden_res0.8', 'leiden_res0.8_colors', 'log1p', 'neighbors', 'pca', 'umap'\n",
       "     obsm: 'X_pca', 'X_umap', 'spatial'\n",
       "     varm: 'PCs'\n",
       "     obsp: 'connectivities', 'distances',\n",
       " 'st_datast_data': View of AnnData object with n_obs × n_vars = 61246 × 4959\n",
       "     obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'n_genes_by_counts', 'n_genes', 'n_counts', 'leiden_res0.8'\n",
       "     var: 'gene_ids', 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'mean', 'std'\n",
       "     uns: 'hvg', 'leiden_res0.8', 'leiden_res0.8_colors', 'log1p', 'neighbors', 'pca', 'umap'\n",
       "     obsm: 'X_pca', 'X_umap', 'spatial'\n",
       "     varm: 'PCs'\n",
       "     obsp: 'connectivities', 'distances',\n",
       " '_i26': 'globals()'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "globals()\n",
    "# del <변수명> -> 삭제"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6a163c8-eb7c-4a70-9eb3-2d7d75d9d96c",
   "metadata": {},
   "source": [
    "# defaulting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "70b77427-554a-46a8-aa71-b29bb6348521",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/dask/dataframe/__init__.py:31: FutureWarning: The legacy Dask DataFrame implementation is deprecated and will be removed in a future version. Set the configuration option `dataframe.query-planning` to `True` or None to enable the new Dask Dataframe implementation and silence this warning.\n",
      "  warnings.warn(\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "# import packages\n",
    "\n",
    "import tissue.main, tissue.downstream\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import os\n",
    "\n",
    "# /home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/dask/dataframe/__init__.py:31: FutureWarning: The legacy Dask DataFrame implementation is deprecated and will be removed in a future version. Set the configuration option `dataframe.query-planning` to `True` or None to enable the new Dask Dataframe implementation and silence this warning.\n",
    "#   warnings.warn(\n",
    "# /home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
    "#   warnings.warn(msg, FutureWarning)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "825b50ae-148a-4121-9781-3e2fc7b1ccac",
   "metadata": {},
   "source": [
    "## loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "e0ba9351-ffce-4706-81c4-192ae063332f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc_data=sc.read_h5ad(\"/data/kjc2/projects/P330.CSA/rds/downsampled_100_re_processed_log1p_harmonized.h5ad\")\n",
    "st_data=sc.read_h5ad(\"/data/kjc2/projects/P330.CSA/rds/250210_CRC_BJM_0050585_Region1_25-04-14-16-53.h5ad\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b1d7de7-7728-4be3-a417-1756f584e557",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## checkout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "55ddfecf-037e-45be-b709-d4d70980c9f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 42649 × 28476\n",
      "    obs: 'dataset', 'medical_condition', 'cancer_type', 'sample_id', 'sample_type', 'tumor_source', 'replicate', 'sample_tissue', 'anatomic_region', 'anatomic_location', 'tumor_stage', 'tumor_stage_TNM', 'tumor_stage_TNM_T', 'tumor_stage_TNM_N', 'tumor_stage_TNM_M', 'tumor_size', 'tumor_dimensions', 'tumor_grade', 'histological_type', 'microsatellite_status', 'mismatch_repair_deficiency_status', 'MLH1_promoter_methylation_status', 'MLH1_status', 'KRAS_status', 'BRAF_status', 'APC_status', 'TP53_status', 'PIK3CA_status', 'SMAD4_status', 'NRAS_status', 'MSH6_status', 'FBXW7_status', 'NOTCH1_status', 'MSH2_status', 'PMS2_status', 'POLE_status', 'ERBB2_status', 'STK11_status', 'HER2_status', 'CTNNB1_status', 'BRAS_status', 'patient_id', 'sex', 'age', 'treatment_status_before_resection', 'treatment_drug', 'treatment_response', 'RECIST', 'platform', 'platform_fine', 'cellranger_version', 'reference_genome', 'matrix_type', 'enrichment_cell_types', 'tissue_cell_state', 'tissue_processing_lab', 'hospital_location', 'country', 'NCBI_BioProject_accession', 'SRA_sample_accession', 'GEO_sample_accession', 'ENA_sample_accession', 'synapse_sample_accession', 'study_id', 'study_doi', 'study_pmid', 'original_obs_names', 'cell_type_coarse_study', 'cell_type_middle_study', 'cell_type_study', 'n_counts', 'n_genes', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_20_genes', 'pct_counts_mito', 'S_score', 'G2M_score', 'phase', 'SOLO_doublet_prob', 'SOLO_singlet_prob', 'SOLO_doublet_status', 'cell_type_predicted', 'cell_type_coarse', 'cell_type_middle', 'cell_type_fine', 'CMS_type', 'immune_infiltration_type', 'is_primary_data', 'suspension_type', 'tissue_type', 'donor_id', 'disease', 'disease_ontology_term_id', 'assay', 'assay_ontology_term_id', 'tissue', 'tissue_ontology_term_id', 'sex_ontology_term_id', 'self_reported_ethnicity', 'self_reported_ethnicity_ontology_term_id', 'organism', 'organism_ontology_term_id', 'development_stage', 'development_stage_ontology_term_id', 'cell_type', 'cell_type_ontology_term_id'\n",
      "    var: 'var_names', 'ensembl', 'Geneid', 'GeneSymbol', 'Chromosome', 'Start', 'End', 'Class', 'Strand', 'Length', 'Version', 'Dataset_25pct_Overlap', 'n_cells', 'n_counts', 'n_cells_by_counts', 'mean_counts', 'total_counts', 'highly_variable', 'original_ensembl_id_with_version', 'base_ensembl_id'\n",
      "    obsm: 'X_scANVI', 'X_scVI', 'X_umap'\n",
      "    layers: 'counts'\n",
      "AnnData object with n_obs × n_vars = 61246 × 1987\n",
      "    obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'n_genes_by_counts', 'n_genes', 'n_counts', 'leiden_res0.8'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'mean', 'std'\n",
      "    uns: 'hvg', 'leiden_res0.8', 'leiden_res0.8_colors', 'log1p', 'neighbors', 'pca', 'umap'\n",
      "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
      "    varm: 'PCs'\n",
      "    obsp: 'connectivities', 'distances'\n",
      "                       var_names          ensembl             Geneid  \\\n",
      "or4f5                      OR4F5  ENSG00000186092  ENSG00000186092.7   \n",
      "ensg00000238009  ENSG00000238009  ENSG00000238009  ENSG00000238009.6   \n",
      "or4f29                    OR4F29  ENSG00000284733  ENSG00000284733.2   \n",
      "ensg00000278791  ENSG00000278791  ENSG00000278791  ENSG00000278791.1   \n",
      "or4f16                    OR4F16  ENSG00000284662  ENSG00000284662.2   \n",
      "...                          ...              ...                ...   \n",
      "rna5-8sn4              RNA5-8SN4  ENSG00000276700  ENSG00000276700.1   \n",
      "loc124907143-1       5_8S_rRNA-8  ENSG00000275757  ENSG00000275757.1   \n",
      "ensg00000276017  ENSG00000276017  ENSG00000276017  ENSG00000276017.1   \n",
      "ensg00000278817  ENSG00000278817  ENSG00000278817  ENSG00000278817.1   \n",
      "ensg00000277196  ENSG00000277196  ENSG00000277196  ENSG00000277196.4   \n",
      "\n",
      "                      GeneSymbol  Chromosome     Start       End  \\\n",
      "or4f5                      OR4F5        chr1   65419.0   71585.0   \n",
      "ensg00000238009  ENSG00000238009        chr1   89295.0  133723.0   \n",
      "or4f29                    OR4F29        chr1  450740.0  451678.0   \n",
      "ensg00000278791  ENSG00000278791        chr1  632325.0  632413.0   \n",
      "or4f16                    OR4F16        chr1  685716.0  686654.0   \n",
      "...                          ...         ...       ...       ...   \n",
      "rna5-8sn4              RNA5-8SN4  KI270733.1  128877.0  129029.0   \n",
      "loc124907143-1         5_8S_rRNA  KI270733.1  173956.0  174108.0   \n",
      "ensg00000276017  ENSG00000276017  KI270734.1   72411.0   74814.0   \n",
      "ensg00000278817  ENSG00000278817  KI270734.1  131494.0  137392.0   \n",
      "ensg00000277196  ENSG00000277196  KI270734.1  138082.0  161852.0   \n",
      "\n",
      "                          Class Strand   Length      Version  \\\n",
      "or4f5            protein_coding      +   6166.0  gencode.v44   \n",
      "ensg00000238009          lncRNA      -  44428.0  gencode.v44   \n",
      "or4f29           protein_coding      -    938.0  gencode.v44   \n",
      "ensg00000278791           miRNA      -     88.0  gencode.v44   \n",
      "or4f16           protein_coding      -    938.0  gencode.v44   \n",
      "...                         ...    ...      ...          ...   \n",
      "rna5-8sn4                  rRNA      +    152.0  gencode.v44   \n",
      "loc124907143-1             rRNA      +    152.0  gencode.v44   \n",
      "ensg00000276017  protein_coding      +   2403.0  gencode.v44   \n",
      "ensg00000278817  protein_coding      +   5898.0  gencode.v44   \n",
      "ensg00000277196  protein_coding      -  23770.0  gencode.v44   \n",
      "\n",
      "                 Dataset_25pct_Overlap  n_cells  n_counts  n_cells_by_counts  \\\n",
      "or4f5                             True      168       693                137   \n",
      "ensg00000238009                   True    14782     23065              11021   \n",
      "or4f29                            True       14       101                 11   \n",
      "ensg00000278791                   True    41004    236173              30119   \n",
      "or4f16                            True       19        61                 14   \n",
      "...                                ...      ...       ...                ...   \n",
      "rna5-8sn4                         True      263       910                147   \n",
      "loc124907143-1                    True      424      1717                250   \n",
      "ensg00000276017                   True      475      2348                429   \n",
      "ensg00000278817                   True    28078     50742              21125   \n",
      "ensg00000277196                   True    14861     47097              12745   \n",
      "\n",
      "                 mean_counts  total_counts  highly_variable  \\\n",
      "or4f5               0.000107         690.0            False   \n",
      "ensg00000238009     0.002041       13184.0            False   \n",
      "or4f29              0.000015          98.0            False   \n",
      "ensg00000278791     0.035371      228456.0            False   \n",
      "or4f16              0.000009          55.0            False   \n",
      "...                      ...           ...              ...   \n",
      "rna5-8sn4           0.000123         794.0            False   \n",
      "loc124907143-1      0.000239        1541.0            False   \n",
      "ensg00000276017     0.001007        6502.0            False   \n",
      "ensg00000278817     0.007658       49459.0            False   \n",
      "ensg00000277196     0.005420       35004.0            False   \n",
      "\n",
      "                original_ensembl_id_with_version  base_ensembl_id  \n",
      "or4f5                            ENSG00000186092  ENSG00000186092  \n",
      "ensg00000238009                  ENSG00000238009  ENSG00000238009  \n",
      "or4f29                           ENSG00000284733  ENSG00000284733  \n",
      "ensg00000278791                  ENSG00000278791  ENSG00000278791  \n",
      "or4f16                           ENSG00000284662  ENSG00000284662  \n",
      "...                                          ...              ...  \n",
      "rna5-8sn4                        ENSG00000276700  ENSG00000276700  \n",
      "loc124907143-1                   ENSG00000275757  ENSG00000275757  \n",
      "ensg00000276017                  ENSG00000276017  ENSG00000276017  \n",
      "ensg00000278817                  ENSG00000278817  ENSG00000278817  \n",
      "ensg00000277196                  ENSG00000277196  ENSG00000277196  \n",
      "\n",
      "[28476 rows x 20 columns]\n",
      "                gene_ids    feature_types   genome  n_cells_by_counts  \\\n",
      "a2ml1    ENSG00000166535  Gene Expression  Unknown                209   \n",
      "abca1    ENSG00000165029  Gene Expression  Unknown               5353   \n",
      "abca3    ENSG00000167972  Gene Expression  Unknown                603   \n",
      "abca4    ENSG00000198691  Gene Expression  Unknown                296   \n",
      "abca7    ENSG00000064687  Gene Expression  Unknown               2190   \n",
      "...                  ...              ...      ...                ...   \n",
      "znf512   ENSG00000243943  Gene Expression  Unknown               5011   \n",
      "znf683   ENSG00000176083  Gene Expression  Unknown               1451   \n",
      "znf804a  ENSG00000170396  Gene Expression  Unknown                181   \n",
      "znf829   ENSG00000185869  Gene Expression  Unknown                348   \n",
      "zscan26  ENSG00000197062  Gene Expression  Unknown               1577   \n",
      "\n",
      "         mean_counts  pct_dropout_by_counts  total_counts  n_cells  \\\n",
      "a2ml1       0.003104              99.691034         210.0      208   \n",
      "abca1       0.108803              92.086629        7360.0     5296   \n",
      "abca3       0.009269              99.108582         627.0      597   \n",
      "abca4       0.004391              99.562421         297.0      293   \n",
      "abca7       0.035184              96.762510        2380.0     2172   \n",
      "...              ...                    ...           ...      ...   \n",
      "znf512      0.081824              92.592209        5535.0     4969   \n",
      "znf683      0.022677              97.854978        1534.0     1421   \n",
      "znf804a     0.002720              99.732427         184.0      179   \n",
      "znf829      0.005322              99.485550         360.0      346   \n",
      "zscan26     0.024082              97.668712        1629.0     1558   \n",
      "\n",
      "         highly_variable  highly_variable_rank     means  variances  \\\n",
      "a2ml1               True                1574.0  0.011040   0.038562   \n",
      "abca1               True                 196.0  0.308923   1.079450   \n",
      "abca3               True                1925.0  0.032374   0.113399   \n",
      "abca4               True                1426.0  0.015814   0.055691   \n",
      "abca7               True                1054.0  0.117631   0.402605   \n",
      "...                  ...                   ...       ...        ...   \n",
      "znf512              True                1029.0  0.267722   0.875266   \n",
      "znf683              True                1095.0  0.077274   0.270892   \n",
      "znf804a             True                1445.0  0.009629   0.033752   \n",
      "znf829              True                1868.0  0.018507   0.064525   \n",
      "zscan26             True                1312.0  0.084308   0.291978   \n",
      "\n",
      "         variances_norm      mean       std  \n",
      "a2ml1          1.011858  0.011040  0.196371  \n",
      "abca1          1.116177  0.308923  1.038966  \n",
      "abca3          1.003484  0.032374  0.336748  \n",
      "abca4          1.016743  0.015814  0.235989  \n",
      "abca7          1.031100  0.117631  0.634512  \n",
      "...                 ...       ...       ...  \n",
      "znf512         1.032390  0.267722  0.935557  \n",
      "znf683         1.029334  0.077274  0.520473  \n",
      "znf804a        1.016041  0.009629  0.183718  \n",
      "znf829         1.004352  0.018507  0.254018  \n",
      "zscan26        1.019857  0.084308  0.540350  \n",
      "\n",
      "[1987 rows x 15 columns]\n"
     ]
    }
   ],
   "source": [
    "print(sc_data)\n",
    "print(st_data)\n",
    "print(sc_data.var)\n",
    "print(st_data.var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0f38b0e5-fc6c-46d5-88db-7305b88df783",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 61246 × 1987\n",
      "    obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'n_genes_by_counts', 'n_genes', 'n_counts', 'leiden_res0.8'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'mean', 'std'\n",
      "    uns: 'hvg', 'leiden_res0.8', 'leiden_res0.8_colors', 'log1p', 'neighbors', 'pca', 'umap'\n",
      "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
      "    varm: 'PCs'\n",
      "    obsp: 'connectivities', 'distances'\n",
      "\n",
      "--- Observations (.obs) ---\n",
      "            x_centroid   y_centroid  transcript_counts  control_probe_counts  \\\n",
      "aaaagckh-1  365.986786  2261.525879                188                     0   \n",
      "aaaaghho-1  386.914917  2254.930664                434                     0   \n",
      "aaaalnmj-1  384.406982  2248.345703                529                     0   \n",
      "aaacelao-1  379.125854  2233.698486                989                     0   \n",
      "aaacjopg-1  311.033691  2303.112793                703                     0   \n",
      "\n",
      "            genomic_control_counts  control_codeword_counts  \\\n",
      "aaaagckh-1                       0                        0   \n",
      "aaaaghho-1                       0                        0   \n",
      "aaaalnmj-1                       0                        0   \n",
      "aaacelao-1                       0                        0   \n",
      "aaacjopg-1                       0                        0   \n",
      "\n",
      "            unassigned_codeword_counts  deprecated_codeword_counts  \\\n",
      "aaaagckh-1                           0                           0   \n",
      "aaaaghho-1                           0                           0   \n",
      "aaaalnmj-1                           0                           0   \n",
      "aaacelao-1                           0                           3   \n",
      "aaacjopg-1                           0                           1   \n",
      "\n",
      "            total_counts   cell_area  nucleus_area  nucleus_count  \\\n",
      "aaaagckh-1         188.0   56.942033     15.669219              2   \n",
      "aaaaghho-1         434.0   67.734377     31.112657              1   \n",
      "aaaalnmj-1         529.0   65.070159     28.583907              1   \n",
      "aaacelao-1         989.0  139.487661     51.162033              2   \n",
      "aaacjopg-1         703.0   71.392034     29.216095              1   \n",
      "\n",
      "                                          segmentation_method  \\\n",
      "aaaagckh-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
      "aaaaghho-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
      "aaaalnmj-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
      "aaacelao-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
      "aaacjopg-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
      "\n",
      "            n_genes_by_counts  n_genes  n_counts leiden_res0.8  \n",
      "aaaagckh-1                153      153     188.0            17  \n",
      "aaaaghho-1                333      333     434.0             6  \n",
      "aaaalnmj-1                408      408     529.0             5  \n",
      "aaacelao-1                615      615     989.0             5  \n",
      "aaacjopg-1                488      488     703.0             5  \n",
      "Columns: Index(['x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts',\n",
      "       'genomic_control_counts', 'control_codeword_counts',\n",
      "       'unassigned_codeword_counts', 'deprecated_codeword_counts',\n",
      "       'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count',\n",
      "       'segmentation_method', 'n_genes_by_counts', 'n_genes', 'n_counts',\n",
      "       'leiden_res0.8'],\n",
      "      dtype='object')\n",
      "\n",
      "--- Variables (.var) ---\n",
      "              gene_ids    feature_types   genome  n_cells_by_counts  \\\n",
      "a2ml1  ENSG00000166535  Gene Expression  Unknown                209   \n",
      "abca1  ENSG00000165029  Gene Expression  Unknown               5353   \n",
      "abca3  ENSG00000167972  Gene Expression  Unknown                603   \n",
      "abca4  ENSG00000198691  Gene Expression  Unknown                296   \n",
      "abca7  ENSG00000064687  Gene Expression  Unknown               2190   \n",
      "\n",
      "       mean_counts  pct_dropout_by_counts  total_counts  n_cells  \\\n",
      "a2ml1     0.003104              99.691034         210.0      208   \n",
      "abca1     0.108803              92.086629        7360.0     5296   \n",
      "abca3     0.009269              99.108582         627.0      597   \n",
      "abca4     0.004391              99.562421         297.0      293   \n",
      "abca7     0.035184              96.762510        2380.0     2172   \n",
      "\n",
      "       highly_variable  highly_variable_rank     means  variances  \\\n",
      "a2ml1             True                1574.0  0.011040   0.038562   \n",
      "abca1             True                 196.0  0.308923   1.079450   \n",
      "abca3             True                1925.0  0.032374   0.113399   \n",
      "abca4             True                1426.0  0.015814   0.055691   \n",
      "abca7             True                1054.0  0.117631   0.402605   \n",
      "\n",
      "       variances_norm      mean       std  \n",
      "a2ml1        1.011858  0.011040  0.196371  \n",
      "abca1        1.116177  0.308923  1.038966  \n",
      "abca3        1.003484  0.032374  0.336748  \n",
      "abca4        1.016743  0.015814  0.235989  \n",
      "abca7        1.031100  0.117631  0.634512  \n",
      "Index name: None\n",
      "Columns: Index(['gene_ids', 'feature_types', 'genome', 'n_cells_by_counts',\n",
      "       'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells',\n",
      "       'highly_variable', 'highly_variable_rank', 'means', 'variances',\n",
      "       'variances_norm', 'mean', 'std'],\n",
      "      dtype='object')\n",
      "\n",
      "--- Expression Matrix (.X) ---\n",
      "Shape: (61246, 1987)\n",
      "Data type: float32\n",
      "Is sparse: False\n",
      "Min value: -1.013006\n",
      "Max value: 10.0\n",
      "\n",
      "--- Layers (.layers) ---\n",
      "KeysView(Layers with keys: )\n",
      "\n",
      "--- Multi-dimensional annotations (.obsm) ---\n",
      "KeysView(AxisArrays with keys: X_pca, X_umap, spatial)\n",
      "Shape of 'spatial': (61246, 2)\n",
      "Shape of 'X_pca': (61246, 50)\n",
      "\n",
      "--- Unstructured annotations (.uns) ---\n",
      "dict_keys(['hvg', 'leiden_res0.8', 'leiden_res0.8_colors', 'log1p', 'neighbors', 'pca', 'umap'])\n",
      "\n",
      "--- Var Names (first 10) ---\n",
      "['a2ml1', 'abca1', 'abca3', 'abca4', 'abca7', 'abcb1', 'abcb4', 'abcc1', 'abcc4', 'abcc9']\n",
      "\n",
      "--- Obs Names (first 10) ---\n",
      "['aaaagckh-1', 'aaaaghho-1', 'aaaalnmj-1', 'aaacelao-1', 'aaacjopg-1', 'aaadfbpe-1', 'aaaecfmd-1', 'aaagfiei-1', 'aaajbacd-1', 'aaakiamn-1']\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# 예시 AnnData 객체 (adata = st_data 또는 sc_data)\n",
    "adata = st_data # 또는 sc_data\n",
    "\n",
    "# 1. 전체 요약 정보 출력 (가장 기본)\n",
    "print(adata)\n",
    "\n",
    "# 2. Jupyter Notebook/Lab 환경에서는 객체 이름만 입력해도 요약 정보가 예쁘게 나옴\n",
    "# adata\n",
    "\n",
    "# 3. 관측치(세포/스팟) 메타데이터 확인 (처음 5개 행)\n",
    "print(\"\\n--- Observations (.obs) ---\")\n",
    "print(adata.obs.head())\n",
    "print(\"Columns:\", adata.obs.columns)\n",
    "\n",
    "# 4. 변수(유전자) 메타데이터 확인 (처음 5개 행)\n",
    "print(\"\\n--- Variables (.var) ---\")\n",
    "print(adata.var.head())\n",
    "print(\"Index name:\", adata.var_names.name) # 보통 유전자 이름\n",
    "print(\"Columns:\", adata.var.columns)\n",
    "\n",
    "# 5. 주 데이터 행렬 (.X) 정보 확인\n",
    "print(\"\\n--- Expression Matrix (.X) ---\")\n",
    "print(\"Shape:\", adata.X.shape)\n",
    "print(\"Data type:\", adata.X.dtype)\n",
    "print(\"Is sparse:\", sp.issparse(adata.X))\n",
    "# print(\"First 5x5 values:\\n\", adata.X[:5, :5].toarray() if sp.issparse(adata.X) else adata.X[:5, :5]) # 실제 값 확인 (클 경우 주의)\n",
    "print(\"Min value:\", adata.X.min())\n",
    "print(\"Max value:\", adata.X.max())\n",
    "\n",
    "# 6. 레이어(Layers) 확인\n",
    "print(\"\\n--- Layers (.layers) ---\")\n",
    "print(adata.layers.keys())\n",
    "if 'counts' in adata.layers:\n",
    "    print(\"Shape of 'counts' layer:\", adata.layers['counts'].shape)\n",
    "    print(\"Data type of 'counts' layer:\", adata.layers['counts'].dtype)\n",
    "\n",
    "# 7. 임베딩 등 다차원 데이터 확인 (.obsm)\n",
    "print(\"\\n--- Multi-dimensional annotations (.obsm) ---\")\n",
    "print(adata.obsm.keys())\n",
    "if 'spatial' in adata.obsm:\n",
    "    print(\"Shape of 'spatial':\", adata.obsm['spatial'].shape)\n",
    "if 'X_pca' in adata.obsm:\n",
    "    print(\"Shape of 'X_pca':\", adata.obsm['X_pca'].shape)\n",
    "\n",
    "# 8. 비정형(Unstructured) 정보 확인 (.uns)\n",
    "print(\"\\n--- Unstructured annotations (.uns) ---\")\n",
    "print(adata.uns.keys())\n",
    "\n",
    "# 9. 유전자 이름 확인 (처음 10개)\n",
    "print(\"\\n--- Var Names (first 10) ---\")\n",
    "print(adata.var_names[:10].to_list())\n",
    "\n",
    "# 10. 관측치 이름(인덱스) 확인 (처음 10개)\n",
    "print(\"\\n--- Obs Names (first 10) ---\")\n",
    "print(adata.obs_names[:10].to_list())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "651839cd-d83c-4ec4-91dc-1f44036df91a",
   "metadata": {},
   "source": [
    "## pipeline pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "450771bb-82fa-4434-8a87-fd5648545f15",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_names: ['ABCA7' 'ABCB1' 'ABCB4' 'ABCC1' 'ABCC2' 'ABCG1' 'ABI3BP' 'ABL2' 'ABTB1'\n",
      " 'ACAP1' 'ACP5' 'ADAM28' 'ADAM33' 'ADAM8' 'ADAMDEC1' 'ADAMTS1' 'ADCYAP1'\n",
      " 'ADGRE2' 'ADGRE5' 'ADM' 'ADRB2' 'AEBP1' 'AFF1' 'AFF2' 'AGR3' 'AICDA'\n",
      " 'AIM2' 'AKT3' 'ALOX5' 'ALPK1' 'ALPL' 'AMIGO3' 'ANGPTL4' 'ANK1' 'ANKHD1'\n",
      " 'ANKRD13A' 'ANKRD13D' 'ANKRD44' 'ANKZF1' 'ANPEP' 'AOAH' 'AP1S3' 'AP5B1'\n",
      " 'APOBEC3G' 'APOL3' 'AQP8' 'AQP9' 'ARC' 'ARG1' 'ARID3B' 'ARL4D' 'ARRB2'\n",
      " 'ARRDC3' 'ASXL1' 'ATF3' 'ATF4' 'ATG10' 'ATG2A' 'ATM' 'ATP2B1' 'ATXN1'\n",
      " 'B3GAT1' 'BACH2' 'BAG3' 'BANK1' 'BATF' 'BAZ2B' 'BBC3' 'BCAS3' 'BCL11A'\n",
      " 'BCL11B' 'BCL2' 'BCL2A1' 'BCL2L11' 'BCL6' 'BCOR' 'BDNF' 'BEX2' 'BFSP2'\n",
      " 'BIRC2' 'BLK' 'BLNK' 'BMP6' 'BNC2' 'BRAF' 'BTAF1' 'BTBD9' 'BTK' 'BTLA'\n",
      " 'C3AR1' 'C5AR1' 'CA4' 'CALB2' 'CALCB' 'CAMP' 'CAPN3' 'CARD11' 'CARD8'\n",
      " 'CASR']\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'ZYX'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'ZYX'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[57], line 18\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m# hold out target gene\u001b[39;00m\n\u001b[1;32m     17\u001b[0m target_gene \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mZYX\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m---> 18\u001b[0m target_expn \u001b[38;5;241m=\u001b[39m \u001b[43mst_data\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtarget_gene\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mX\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[1;32m     19\u001b[0m st_data \u001b[38;5;241m=\u001b[39m st_data[:, [gene \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m gene_names \u001b[38;5;28;01mif\u001b[39;00m gene \u001b[38;5;241m!=\u001b[39m target_gene]]\u001b[38;5;241m.\u001b[39mcopy()\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/anndata/_core/anndata.py:1011\u001b[0m, in \u001b[0;36mAnnData.__getitem__\u001b[0;34m(self, index)\u001b[0m\n\u001b[1;32m   1009\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m__getitem__\u001b[39m(\u001b[38;5;28mself\u001b[39m, index: Index) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m AnnData:\n\u001b[1;32m   1010\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Returns a sliced view of the object.\"\"\"\u001b[39;00m\n\u001b[0;32m-> 1011\u001b[0m     oidx, vidx \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_normalize_indices\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindex\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1012\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m AnnData(\u001b[38;5;28mself\u001b[39m, oidx\u001b[38;5;241m=\u001b[39moidx, vidx\u001b[38;5;241m=\u001b[39mvidx, asview\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/anndata/_core/anndata.py:992\u001b[0m, in \u001b[0;36mAnnData._normalize_indices\u001b[0;34m(self, index)\u001b[0m\n\u001b[1;32m    991\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_normalize_indices\u001b[39m(\u001b[38;5;28mself\u001b[39m, index: Index \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mtuple\u001b[39m[\u001b[38;5;28mslice\u001b[39m, \u001b[38;5;28mslice\u001b[39m]:\n\u001b[0;32m--> 992\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_normalize_indices\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobs_names\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvar_names\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/anndata/_core/index.py:33\u001b[0m, in \u001b[0;36m_normalize_indices\u001b[0;34m(index, names0, names1)\u001b[0m\n\u001b[1;32m     31\u001b[0m ax0, ax1 \u001b[38;5;241m=\u001b[39m unpack_index(index)\n\u001b[1;32m     32\u001b[0m ax0 \u001b[38;5;241m=\u001b[39m _normalize_index(ax0, names0)\n\u001b[0;32m---> 33\u001b[0m ax1 \u001b[38;5;241m=\u001b[39m \u001b[43m_normalize_index\u001b[49m\u001b[43m(\u001b[49m\u001b[43max1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnames1\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     34\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m ax0, ax1\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/anndata/_core/index.py:70\u001b[0m, in \u001b[0;36m_normalize_index\u001b[0;34m(indexer, index)\u001b[0m\n\u001b[1;32m     68\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m indexer\n\u001b[1;32m     69\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(indexer, \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m---> 70\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# int\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\n\u001b[1;32m     72\u001b[0m     indexer, Sequence \u001b[38;5;241m|\u001b[39m np\u001b[38;5;241m.\u001b[39mndarray \u001b[38;5;241m|\u001b[39m pd\u001b[38;5;241m.\u001b[39mIndex \u001b[38;5;241m|\u001b[39m spmatrix \u001b[38;5;241m|\u001b[39m np\u001b[38;5;241m.\u001b[39mmatrix \u001b[38;5;241m|\u001b[39m SpArray\n\u001b[1;32m     73\u001b[0m ):\n\u001b[1;32m     74\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mshape\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m (\n\u001b[1;32m     75\u001b[0m         (indexer\u001b[38;5;241m.\u001b[39mshape \u001b[38;5;241m==\u001b[39m (index\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m], \u001b[38;5;241m1\u001b[39m))\n\u001b[1;32m     76\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m (indexer\u001b[38;5;241m.\u001b[39mshape \u001b[38;5;241m==\u001b[39m (\u001b[38;5;241m1\u001b[39m, index\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m]))\n\u001b[1;32m     77\u001b[0m     ):\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'ZYX'"
     ]
    }
   ],
   "source": [
    "# pipeline pre-processing\n",
    "\n",
    "# # make genes lowercase\n",
    "# adata.var_names = [x.lower() for x in adata.var_names]\n",
    "# RNAseq_adata.var_names = [x.lower() for x in RNAseq_adata.var_names]\n",
    "\n",
    "# preprocess RNAseq data\n",
    "# tissue.main.preprocess_data(RNAseq_adata, standardize=False, normalize=True)\n",
    "tissue.main.preprocess_data(sc_data, standardize=False, normalize=False)\n",
    "\n",
    "# subset spatial data into shared genes\n",
    "gene_names = np.intersect1d(st_data.var_names, sc_data.var_names)\n",
    "print(f'gene_names: {gene_names[1:100]}')\n",
    "st_data = st_data[:, gene_names].copy()\n",
    "\n",
    "# hold out target gene\n",
    "target_gene = \"CASR\"\n",
    "target_expn = st_data[:, target_gene].X.copy()\n",
    "st_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "92e952ff-4977-4385-a1a8-9c9592acd30d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hold out target gene\n",
    "target_gene = \"CASR\"\n",
    "target_expn = st_data[:, target_gene].X.copy()\n",
    "st_data = st_data[:, [gene for gene in gene_names if gene != target_gene]].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a1b3422d-d31c-4f98-8b78-b323dc1a9596",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_names: ['A2ML1' 'AAMP' 'AAR2' ... 'ZUP1' 'ZYG11B' 'ZYX']\n"
     ]
    }
   ],
   "source": [
    "print(f'gene_names: {gene_names}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80f06618-a029-4530-afb6-b78fe582345f",
   "metadata": {},
   "source": [
    "# data checkout"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb207792-b79a-4744-9d4e-f142e513e166",
   "metadata": {},
   "source": [
    "## shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "39b5a919-d7fe-4708-bfdc-59d11373adb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(42648, 3662)\n",
      "(62941, 963)\n"
     ]
    }
   ],
   "source": [
    "print(sc_data.shape)\n",
    "print(st_data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "02d87e4d-3d09-4cef-869d-12743971c350",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 42648 × 3662\n",
      "    obs: 'dataset', 'medical_condition', 'cancer_type', 'sample_id', 'sample_type', 'tumor_source', 'replicate', 'sample_tissue', 'anatomic_region', 'anatomic_location', 'tumor_stage', 'tumor_stage_TNM', 'tumor_stage_TNM_T', 'tumor_stage_TNM_N', 'tumor_stage_TNM_M', 'tumor_size', 'tumor_dimensions', 'tumor_grade', 'histological_type', 'microsatellite_status', 'mismatch_repair_deficiency_status', 'MLH1_promoter_methylation_status', 'MLH1_status', 'KRAS_status', 'BRAF_status', 'APC_status', 'TP53_status', 'PIK3CA_status', 'SMAD4_status', 'NRAS_status', 'MSH6_status', 'FBXW7_status', 'NOTCH1_status', 'MSH2_status', 'PMS2_status', 'POLE_status', 'ERBB2_status', 'STK11_status', 'HER2_status', 'CTNNB1_status', 'BRAS_status', 'patient_id', 'sex', 'age', 'treatment_status_before_resection', 'treatment_drug', 'treatment_response', 'RECIST', 'platform', 'platform_fine', 'cellranger_version', 'reference_genome', 'matrix_type', 'enrichment_cell_types', 'tissue_cell_state', 'tissue_processing_lab', 'hospital_location', 'country', 'NCBI_BioProject_accession', 'SRA_sample_accession', 'GEO_sample_accession', 'ENA_sample_accession', 'synapse_sample_accession', 'study_id', 'study_doi', 'study_pmid', 'original_obs_names', 'cell_type_coarse_study', 'cell_type_middle_study', 'cell_type_study', 'n_counts', 'n_genes', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_20_genes', 'pct_counts_mito', 'S_score', 'G2M_score', 'phase', 'SOLO_doublet_prob', 'SOLO_singlet_prob', 'SOLO_doublet_status', 'cell_type_predicted', 'cell_type_coarse', 'cell_type_middle', 'cell_type_fine', 'CMS_type', 'immune_infiltration_type', 'is_primary_data', 'suspension_type', 'tissue_type', 'donor_id', 'disease', 'disease_ontology_term_id', 'assay', 'assay_ontology_term_id', 'tissue', 'tissue_ontology_term_id', 'sex_ontology_term_id', 'self_reported_ethnicity', 'self_reported_ethnicity_ontology_term_id', 'organism', 'organism_ontology_term_id', 'development_stage', 'development_stage_ontology_term_id', 'cell_type', 'cell_type_ontology_term_id', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'leiden'\n",
      "    var: 'var_names', 'ensembl', 'Geneid', 'GeneSymbol', 'Chromosome', 'Start', 'End', 'Class', 'Strand', 'Length', 'Version', 'Dataset_25pct_Overlap', 'n_cells', 'n_counts', 'n_cells_by_counts', 'mean_counts', 'total_counts', 'highly_variable', 'original_ensembl_id_with_version', 'base_ensembl_id', 'mt', 'ribo', 'hb', 'log1p_mean_counts', 'pct_dropout_by_counts', 'log1p_total_counts', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
      "    uns: 'cell_type_colors', 'dataset_colors', 'hvg', 'leiden', 'leiden_colors', 'log1p', 'neighbors', 'pca', 'rank_genes_groups', 'umap'\n",
      "    obsm: 'X_pca', 'X_pca_harmony', 'X_scANVI', 'X_scVI', 'X_umap'\n",
      "    varm: 'PCs'\n",
      "    layers: 'counts'\n",
      "    obsp: 'connectivities', 'distances'\n",
      "AnnData object with n_obs × n_vars = 62941 × 963\n",
      "    obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'n_genes_by_counts', 'n_genes', 'n_counts', 'leiden_res0.8'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'mean', 'std'\n",
      "    uns: 'hvg', 'leiden_res0.8', 'leiden_res0.8_colors', 'log1p', 'neighbors', 'pca', 'umap'\n",
      "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
      "    varm: 'PCs'\n",
      "    obsp: 'connectivities', 'distances'\n"
     ]
    }
   ],
   "source": [
    "print(sc_data)\n",
    "print(st_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69cc4e48-306e-4aa6-87bc-4cf64bc68cfb",
   "metadata": {},
   "source": [
    "## nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0bb91ec0-689c-445a-8511-1a8183b959e0",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "x_centroid                    0\n",
       "y_centroid                    0\n",
       "transcript_counts             0\n",
       "control_probe_counts          0\n",
       "genomic_control_counts        0\n",
       "control_codeword_counts       0\n",
       "unassigned_codeword_counts    0\n",
       "deprecated_codeword_counts    0\n",
       "total_counts                  0\n",
       "cell_area                     0\n",
       "nucleus_area                  0\n",
       "nucleus_count                 0\n",
       "segmentation_method           0\n",
       "n_genes_by_counts             0\n",
       "n_genes                       0\n",
       "n_counts                      0\n",
       "leiden_res0.8                 0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "st_data.obs.isna().sum()  # 각 열별 NA 개수"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b31b916d-f422-4521-bedb-6e48868b183c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_ids                 0\n",
       "feature_types            0\n",
       "genome                   0\n",
       "n_cells_by_counts        0\n",
       "mean_counts              0\n",
       "pct_dropout_by_counts    0\n",
       "total_counts             0\n",
       "n_cells                  0\n",
       "highly_variable          0\n",
       "highly_variable_rank     0\n",
       "means                    0\n",
       "variances                0\n",
       "variances_norm           0\n",
       "mean                     0\n",
       "std                      0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "st_data.var.isna().sum() # 각 열별 NA 개수"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1032097d-1603-4ca6-a5df-a05801f74c0f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x_centroid</th>\n",
       "      <th>y_centroid</th>\n",
       "      <th>transcript_counts</th>\n",
       "      <th>control_probe_counts</th>\n",
       "      <th>genomic_control_counts</th>\n",
       "      <th>control_codeword_counts</th>\n",
       "      <th>unassigned_codeword_counts</th>\n",
       "      <th>deprecated_codeword_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>cell_area</th>\n",
       "      <th>nucleus_area</th>\n",
       "      <th>nucleus_count</th>\n",
       "      <th>segmentation_method</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>leiden_res0.8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aacmaokk-1</th>\n",
       "      <td>405.386169</td>\n",
       "      <td>2441.927979</td>\n",
       "      <td>459</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>459.0</td>\n",
       "      <td>40.730939</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Segmented by boundary stain (ATP1A1+CD45+E-Cad...</td>\n",
       "      <td>362</td>\n",
       "      <td>362</td>\n",
       "      <td>459.0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aadaljne-1</th>\n",
       "      <td>368.016602</td>\n",
       "      <td>2380.937744</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>20.681563</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Segmented by boundary stain (ATP1A1+CD45+E-Cad...</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>57.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaefagpo-1</th>\n",
       "      <td>351.166016</td>\n",
       "      <td>3623.459229</td>\n",
       "      <td>256</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>256.0</td>\n",
       "      <td>26.551876</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Segmented by boundary stain (ATP1A1+CD45+E-Cad...</td>\n",
       "      <td>229</td>\n",
       "      <td>229</td>\n",
       "      <td>256.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaehngjp-1</th>\n",
       "      <td>337.040192</td>\n",
       "      <td>3655.487793</td>\n",
       "      <td>266</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>266.0</td>\n",
       "      <td>23.390938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Segmented by boundary stain (ATP1A1+CD45+E-Cad...</td>\n",
       "      <td>220</td>\n",
       "      <td>220</td>\n",
       "      <td>266.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaekadjc-1</th>\n",
       "      <td>344.637512</td>\n",
       "      <td>3629.710938</td>\n",
       "      <td>348</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>29.577345</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Segmented by boundary stain (ATP1A1+CD45+E-Cad...</td>\n",
       "      <td>295</td>\n",
       "      <td>295</td>\n",
       "      <td>348.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>djhekopl-1</th>\n",
       "      <td>545.984497</td>\n",
       "      <td>4863.115234</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>460.0</td>\n",
       "      <td>30.209532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Segmented by boundary stain (ATP1A1+CD45+E-Cad...</td>\n",
       "      <td>339</td>\n",
       "      <td>339</td>\n",
       "      <td>460.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>djhjicnp-1</th>\n",
       "      <td>551.421509</td>\n",
       "      <td>4860.001465</td>\n",
       "      <td>478</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>478.0</td>\n",
       "      <td>28.403282</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Segmented by boundary stain (ATP1A1+CD45+E-Cad...</td>\n",
       "      <td>366</td>\n",
       "      <td>366</td>\n",
       "      <td>478.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>djhkeoke-1</th>\n",
       "      <td>555.476868</td>\n",
       "      <td>4852.318848</td>\n",
       "      <td>461</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>461.0</td>\n",
       "      <td>38.157033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Segmented by boundary stain (ATP1A1+CD45+E-Cad...</td>\n",
       "      <td>331</td>\n",
       "      <td>331</td>\n",
       "      <td>461.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>djigmoec-1</th>\n",
       "      <td>523.969360</td>\n",
       "      <td>4901.240723</td>\n",
       "      <td>483</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>483.0</td>\n",
       "      <td>28.448439</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Segmented by boundary stain (ATP1A1+CD45+E-Cad...</td>\n",
       "      <td>344</td>\n",
       "      <td>344</td>\n",
       "      <td>483.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>djjchnhc-1</th>\n",
       "      <td>567.563660</td>\n",
       "      <td>4861.365723</td>\n",
       "      <td>491</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>491.0</td>\n",
       "      <td>31.338439</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Segmented by boundary stain (ATP1A1+CD45+E-Cad...</td>\n",
       "      <td>359</td>\n",
       "      <td>359</td>\n",
       "      <td>491.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1695 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            x_centroid   y_centroid  transcript_counts  control_probe_counts  \\\n",
       "aacmaokk-1  405.386169  2441.927979                459                     0   \n",
       "aadaljne-1  368.016602  2380.937744                 57                     0   \n",
       "aaefagpo-1  351.166016  3623.459229                256                     0   \n",
       "aaehngjp-1  337.040192  3655.487793                266                     0   \n",
       "aaekadjc-1  344.637512  3629.710938                348                     0   \n",
       "...                ...          ...                ...                   ...   \n",
       "djhekopl-1  545.984497  4863.115234                460                     0   \n",
       "djhjicnp-1  551.421509  4860.001465                478                     0   \n",
       "djhkeoke-1  555.476868  4852.318848                461                     0   \n",
       "djigmoec-1  523.969360  4901.240723                483                     0   \n",
       "djjchnhc-1  567.563660  4861.365723                491                     0   \n",
       "\n",
       "            genomic_control_counts  control_codeword_counts  \\\n",
       "aacmaokk-1                       0                        0   \n",
       "aadaljne-1                       0                        0   \n",
       "aaefagpo-1                       0                        0   \n",
       "aaehngjp-1                       0                        0   \n",
       "aaekadjc-1                       0                        0   \n",
       "...                            ...                      ...   \n",
       "djhekopl-1                       0                        0   \n",
       "djhjicnp-1                       0                        0   \n",
       "djhkeoke-1                       0                        0   \n",
       "djigmoec-1                       0                        0   \n",
       "djjchnhc-1                       0                        0   \n",
       "\n",
       "            unassigned_codeword_counts  deprecated_codeword_counts  \\\n",
       "aacmaokk-1                           0                           1   \n",
       "aadaljne-1                           0                           0   \n",
       "aaefagpo-1                           0                           2   \n",
       "aaehngjp-1                           0                           0   \n",
       "aaekadjc-1                           0                           0   \n",
       "...                                ...                         ...   \n",
       "djhekopl-1                           0                           0   \n",
       "djhjicnp-1                           0                           0   \n",
       "djhkeoke-1                           0                           0   \n",
       "djigmoec-1                           0                           1   \n",
       "djjchnhc-1                           0                           0   \n",
       "\n",
       "            total_counts  cell_area  nucleus_area  nucleus_count  \\\n",
       "aacmaokk-1         459.0  40.730939           NaN              0   \n",
       "aadaljne-1          57.0  20.681563           NaN              0   \n",
       "aaefagpo-1         256.0  26.551876           NaN              0   \n",
       "aaehngjp-1         266.0  23.390938           NaN              0   \n",
       "aaekadjc-1         348.0  29.577345           NaN              0   \n",
       "...                  ...        ...           ...            ...   \n",
       "djhekopl-1         460.0  30.209532           NaN              0   \n",
       "djhjicnp-1         478.0  28.403282           NaN              0   \n",
       "djhkeoke-1         461.0  38.157033           NaN              0   \n",
       "djigmoec-1         483.0  28.448439           NaN              0   \n",
       "djjchnhc-1         491.0  31.338439           NaN              0   \n",
       "\n",
       "                                          segmentation_method  \\\n",
       "aacmaokk-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "aadaljne-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "aaefagpo-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "aaehngjp-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "aaekadjc-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "...                                                       ...   \n",
       "djhekopl-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "djhjicnp-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "djhkeoke-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "djigmoec-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "djjchnhc-1  Segmented by boundary stain (ATP1A1+CD45+E-Cad...   \n",
       "\n",
       "            n_genes_by_counts  n_genes  n_counts leiden_res0.8  \n",
       "aacmaokk-1                362      362     459.0             5  \n",
       "aadaljne-1                 50       50      57.0             2  \n",
       "aaefagpo-1                229      229     256.0             2  \n",
       "aaehngjp-1                220      220     266.0             2  \n",
       "aaekadjc-1                295      295     348.0             2  \n",
       "...                       ...      ...       ...           ...  \n",
       "djhekopl-1                339      339     460.0             0  \n",
       "djhjicnp-1                366      366     478.0             0  \n",
       "djhkeoke-1                331      331     461.0             0  \n",
       "djigmoec-1                344      344     483.0             0  \n",
       "djjchnhc-1                359      359     491.0             0  \n",
       "\n",
       "[1695 rows x 17 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "st_data.obs[st_data.obs.isna().any(axis=1)] #NA가 있는 행들 보기\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "452bedc5-0da8-4e11-8484-2676f57e83d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "st_data.var[st_data.var.isna().any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7009fdf8-caad-44a1-b542-50491cc0c3c1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'counts'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[19], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m np\u001b[38;5;241m.\u001b[39misnan(\u001b[43mst_data\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlayers\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcounts\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m)\u001b[38;5;241m.\u001b[39msum() \u001b[38;5;66;03m#특정 layers의 NA\u001b[39;00m\n\u001b[1;32m      2\u001b[0m st_data\u001b[38;5;241m.\u001b[39mobsm[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mX_pca\u001b[39m\u001b[38;5;124m'\u001b[39m][np\u001b[38;5;241m.\u001b[39misnan(st_data\u001b[38;5;241m.\u001b[39mobsm[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mX_pca\u001b[39m\u001b[38;5;124m'\u001b[39m])\u001b[38;5;241m.\u001b[39many(axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)] \u001b[38;5;66;03m#.obsm, .varm 등에서도.\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/anndata/_core/aligned_mapping.py:211\u001b[0m, in \u001b[0;36mAlignedActual.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    210\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m__getitem__\u001b[39m(\u001b[38;5;28mself\u001b[39m, key: \u001b[38;5;28mstr\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Value:\n\u001b[0;32m--> 211\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_data\u001b[49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'counts'"
     ]
    }
   ],
   "source": [
    "np.isnan(st_data.layers[\"counts\"]).sum() #특정 layers의 NA\n",
    "st_data.obsm['X_pca'][np.isnan(st_data.obsm['X_pca']).any(axis=1)] #.obsm, .varm 등에서도.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ed6a41-128c-4bde-b38f-469651cfbd45",
   "metadata": {},
   "source": [
    "# 결측치 다루기"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23ec4c95-38f7-463c-a414-c9ab0a2cfda8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## 메타데이터"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d878aa1f-4f60-46d1-b0d3-4b952b98be75",
   "metadata": {},
   "source": [
    "### fillna(0) fillna(median_area)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36747340-6384-4a92-a0cd-d3b1c6633b4b",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "st_data_original=st_data\n",
    "st_data.obs['highly_variable_rank'] = st_data.obs['highly_variable_rank'].fillna(0)\n",
    "median_area = st_data.var['nucleus_area'].median()\n",
    "st_data.var['nucleus_area'] = st_data.var['nucleus_area'].fillna(median_area)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "239adbbb-2bd8-4e6c-931b-d476999743de",
   "metadata": {},
   "source": [
    "### 제거"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ca1a930-f4be-4e0c-9abe-34efdcfa9c08",
   "metadata": {},
   "source": [
    "#### na만 제거"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "16d479f2-3d2a-4ac7-8e37-76071eb9b487",
   "metadata": {},
   "outputs": [],
   "source": [
    "st_data_original=st_data\n",
    "st_data= st_data[~st_data.obs.isna().any(axis=1), :]\n",
    "st_data = st_data[:, ~st_data.var.isna().any(axis=1)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e3e3afaf-58e2-465a-bd76-370723bcc887",
   "metadata": {},
   "outputs": [],
   "source": [
    "st_data= st_data[~st_data.obs.isna().any(axis=1), :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c200a35-8b7c-4ad0-aa59-62d78077d38b",
   "metadata": {},
   "source": [
    "#### 해당 메타데이터 제거"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4efe5c61-d196-4200-befa-e7259a0c2c75",
   "metadata": {},
   "outputs": [],
   "source": [
    "st_data.obs = st_data.obs.drop(columns=['highly_variable_rank'])\n",
    "st_data.var = st_data.var.drop(columns=['nucleus_area'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f7ee9047-1133-4eb2-9e63-3bb5bb71f725",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "View of AnnData object with n_obs × n_vars = 61246 × 4959\n",
      "    obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'n_genes_by_counts', 'n_genes', 'n_counts', 'leiden_res0.8'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'mean', 'std'\n",
      "    uns: 'hvg', 'leiden_res0.8', 'leiden_res0.8_colors', 'log1p', 'neighbors', 'pca', 'umap'\n",
      "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
      "    varm: 'PCs'\n",
      "    obsp: 'connectivities', 'distances'\n"
     ]
    }
   ],
   "source": [
    "print(st_data_original)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f25826d-b946-4ee5-8421-d2e7642d7fb6",
   "metadata": {},
   "source": [
    "## data matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "5ea5ba45-8f22-474a-9b2e-2f03d5011312",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NaN in st_data.X: False\n",
      "NaN count in sc_data.X: 0\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "# NaN이 있는지\n",
    "\n",
    "print(\"NaN in st_data.X:\", np.isnan(st_data.X).any())\n",
    "# print(\"NaN in sc_data.X:\", np.isnan(sc_data.X).any())\n",
    "# NaN in st_data.X: False\n",
    "# ---------------------------------------------------------------------------\n",
    "# TypeError                                 Traceback (most recent call last)\n",
    "# Cell In[39], line 4\n",
    "#       1 import numpy as np\n",
    "#       3 print(\"NaN in st_data.X:\", np.isnan(st_data.X).any())\n",
    "# ----> 4 print(\"NaN in sc_data.X:\", np.isnan(sc_data.X).any())\n",
    "#       6 # 만약 .X가 매우 크다면 .toarray()는 메모리 부족을 유발할 수 있습니다.\n",
    "#       7 # print(\"NaN count in st_data.X:\", np.isnan(st_data.X.toarray()).sum())\n",
    "#       8 # print(\"NaN count in sc_data.X:\", np.isnan(sc_data.X.toarray()).sum())\n",
    "\n",
    "# TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''\n",
    "\n",
    "# 만약 .X가 매우 크다면 .toarray()는 메모리 부족을 유발할 수 있습니다.\n",
    "# print(\"NaN count in st_data.X:\", np.isnan(st_data.X.toarray()).sum())\n",
    "print(\"NaN count in sc_data.X:\", np.isnan(sc_data.X.toarray()).sum())\n",
    "# ---------------------------------------------------------------------------\n",
    "# AttributeError                            Traceback (most recent call last)\n",
    "# Cell In[40], line 7\n",
    "#       1 import numpy as np\n",
    "#       3 # print(\"NaN in st_data.X:\", np.isnan(st_data.X).any())\n",
    "#       4 # print(\"NaN in sc_data.X:\", np.isnan(sc_data.X).any())\n",
    "#       5 \n",
    "#       6 # 만약 .X가 매우 크다면 .toarray()는 메모리 부족을 유발할 수 있습니다.\n",
    "# ----> 7 print(\"NaN count in st_data.X:\", np.isnan(st_data.X.toarray()).sum())\n",
    "#       8 print(\"NaN count in sc_data.X:\", np.isnan(sc_data.X.toarray()).sum())\n",
    "\n",
    "# AttributeError: 'numpy.ndarray' object has no attribute 'toarray'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6229a6f5-9e1a-49cf-9e26-1c17ada68197",
   "metadata": {},
   "source": [
    "### 데이터 타입. TISSUE 사용을 위해서는 integer가 있으면 안 된다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "25b9e93e-c7bb-44b4-a784-0dbd4aa2f6ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "st_data.X dtype: float32\n",
      "sc_data.X dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print(\"st_data.X dtype:\", st_data.X.dtype)\n",
    "print(\"sc_data.X dtype:\", sc_data.X.dtype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "aa5a6e0e-8a2a-48bc-b3a9-6a8fd56bf626",
   "metadata": {},
   "outputs": [],
   "source": [
    "st_data.X = st_data.X.astype(np.float32)\n",
    "sc_data.X = sc_data.X.astype(np.float32)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6314acad-f9f1-4ebf-ae50-6d026175466e",
   "metadata": {},
   "source": [
    "### 제거"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e6cd3fd-6293-435f-a214-294dfb9ac773",
   "metadata": {},
   "source": [
    "#### 0으로"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f831a3c3-69ac-4f07-a20a-b8e943683ad4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b55c0f20-0285-4fc4-9baa-b4b733bef7b2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### 해당 열(유전자) 제거"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6e15247-a916-4dd2-aff9-52eb2323a070",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# st_data에서 NaN 포함 유전자 제거\n",
    "if hasattr(st_data.X, 'data'): # 희소 행렬\n",
    "     genes_to_keep_st = ~np.isnan(st_data.X.A).any(axis=0) # .A로 밀집 변환 후 확인 (메모리 주의)\n",
    "else: # 밀집 행렬\n",
    "     genes_to_keep_st = ~np.isnan(st_data.X).any(axis=0)\n",
    "print(f\"Removing {np.sum(~genes_to_keep_st)} genes with NaN from st_data\")\n",
    "st_data = st_data[:, genes_to_keep_st].copy()\n",
    "\n",
    "# sc_data에서 NaN 포함 유전자 제거 (동일 방식)\n",
    "if hasattr(sc_data.X, 'data'): # 희소 행렬\n",
    "     genes_to_keep_sc = ~np.isnan(sc_data.X.A).any(axis=0) # .A로 밀집 변환 후 확인 (메모리 주의)\n",
    "else: # 밀집 행렬\n",
    "     genes_to_keep_sc = ~np.isnan(sc_data.X).any(axis=0)\n",
    "print(f\"Removing {np.sum(~genes_to_keep_sc)} genes with NaN from sc_data\")\n",
    "sc_data = sc_data[:, genes_to_keep_sc].copy()\n",
    "\n",
    "# target_gene이 제거되지 않았는지 확인 후 TISSUE 실행\n",
    "if target_gene in st_data.var_names and target_gene in sc_data.var_names:\n",
    "     # tissue.main.predict_gene_expression(...)\n",
    "     pass\n",
    "else:\n",
    "     print(f\"Target gene {target_gene} was removed due to NaN values.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeacf89c-502c-4cc4-a27f-0dc09c8a9e17",
   "metadata": {},
   "source": [
    "#### 해당 셀 제거: 공간 데이터(st_data)에서는 공간적 위치 정보를 잃을 수 있으므로 신중해야 합니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19c7f7b8-5321-42c8-ac2c-3591f0de92ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# st_data에서 NaN 포함 셀/스팟 제거\n",
    "if hasattr(st_data.X, 'data'): # 희소 행렬\n",
    "    cells_to_keep_st = ~np.isnan(st_data.X.A).any(axis=1) # .A 사용 주의\n",
    "else: # 밀집 행렬\n",
    "    cells_to_keep_st = ~np.isnan(st_data.X).any(axis=1)\n",
    "print(f\"Removing {np.sum(~cells_to_keep_st)} spots/cells with NaN from st_data\")\n",
    "st_data = st_data[cells_to_keep_st, :].copy()\n",
    "\n",
    "# sc_data에서 NaN 포함 셀 제거 (동일 방식)\n",
    "if hasattr(sc_data.X, 'data'): # 희소 행렬\n",
    "    cells_to_keep_sc = ~np.isnan(sc_data.X.A).any(axis=1) # .A 사용 주의\n",
    "else: # 밀집 행렬\n",
    "    cells_to_keep_sc = ~np.isnan(sc_data.X).any(axis=1)\n",
    "print(f\"Removing {np.sum(~cells_to_keep_sc)} cells with NaN from sc_data\")\n",
    "sc_data = sc_data[cells_to_keep_sc, :].copy()\n",
    "\n",
    "# TISSUE 함수 다시 실행\n",
    "# tissue.main.predict_gene_expression(...)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "136a7fbc-d659-410b-a456-9794c9412141",
   "metadata": {},
   "source": [
    "#### 데이터 형식 일치시키기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd857664-e062-451e-9f4e-571d168fef40",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scipy.sparse as sp\n",
    "\n",
    "# sc_data.X의 데이터 타입을 float32로 변경\n",
    "if isinstance(sc_data.X, np.ndarray):\n",
    "    sc_data.X = sc_data.X.astype(np.float32)\n",
    "elif sp.issparse(sc_data.X):\n",
    "    # 희소 행렬의 경우, data 속성의 타입만 변경\n",
    "    sc_data.X = sc_data.X.astype(np.float32)\n",
    "else:\n",
    "    # 다른 타입일 경우 밀집 배열로 변환 후 변경 (메모리 주의)\n",
    "    print(\"Warning: Converting sc_data.X to dense array for dtype change.\")\n",
    "    sc_data.X = sc_data.X.toarray().astype(np.float32)\n",
    "\n",
    "# st_data.X가 이미 float32가 아니라면 변환 (이미 하셨으므로 확인용)\n",
    "if st_data.X.dtype != np.float32:\n",
    "     if isinstance(st_data.X, np.ndarray):\n",
    "        st_data.X = st_data.X.astype(np.float32)\n",
    "     elif sp.issparse(st_data.X):\n",
    "        st_data.X = st_data.X.astype(np.float32)\n",
    "     else:\n",
    "        print(\"Warning: Converting st_data.X to dense array for dtype change.\")\n",
    "        st_data.X = st_data.X.toarray().astype(np.float32)\n",
    "\n",
    "\n",
    "print(\"After conversion:\")\n",
    "print(\"st_data.X dtype:\", st_data.X.dtype)\n",
    "print(\"sc_data.X dtype:\", sc_data.X.dtype)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0df9024-2003-4c1c-8fb7-45d4e9a54434",
   "metadata": {},
   "source": [
    "#### NaN 재확인 - 형식 고려"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "2f7d4dfa-cea8-4df7-98e5-dfa1db464d28",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Checking st_data (<class 'numpy.ndarray'>) ---\n",
      "NaN in st_data.X: False (count: 0)\n",
      "Inf in st_data.X: False (count: 0)\n",
      "--- Checking sc_data (<class 'scipy.sparse._csr.csr_matrix'>) ---\n",
      "NaN in sc_data.X.data: False (count: 0)\n",
      "Inf in sc_data.X.data: False (count: 0)\n",
      "NaN in sc_data.layers['counts'].data: False (count: 0)\n",
      "Inf in sc_data.layers['counts'].data: False (count: 0)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import scipy.sparse as sp\n",
    "\n",
    "def check_nan_inf(adata, name):\n",
    "    print(f\"--- Checking {name} ({type(adata.X)}) ---\")\n",
    "    has_nan = False\n",
    "    has_inf = False\n",
    "\n",
    "    if sp.issparse(adata.X):\n",
    "        # 희소 행렬: data 속성에서 NaN/inf 확인\n",
    "        has_nan = np.isnan(adata.X.data).any()\n",
    "        has_inf = np.isinf(adata.X.data).any()\n",
    "        print(f\"NaN in {name}.X.data: {has_nan} (count: {np.isnan(adata.X.data).sum()})\")\n",
    "        print(f\"Inf in {name}.X.data: {has_inf} (count: {np.isinf(adata.X.data).sum()})\")\n",
    "    elif isinstance(adata.X, np.ndarray):\n",
    "        # 밀집 행렬: 전체 행렬에서 NaN/inf 확인\n",
    "        has_nan = np.isnan(adata.X).any()\n",
    "        has_inf = np.isinf(adata.X).any()\n",
    "        print(f\"NaN in {name}.X: {has_nan} (count: {np.isnan(adata.X).sum()})\")\n",
    "        print(f\"Inf in {name}.X: {has_inf} (count: {np.isinf(adata.X).sum()})\")\n",
    "    else:\n",
    "        print(f\"Warning: Unsupported type for {name}.X: {type(adata.X)}. Skipping NaN/Inf check.\")\n",
    "\n",
    "    # 레이어 확인 (sc_data에 'counts' 레이어가 있으므로 확인)\n",
    "    if 'counts' in adata.layers:\n",
    "        layer_data = adata.layers['counts']\n",
    "        layer_has_nan = False\n",
    "        layer_has_inf = False\n",
    "        if sp.issparse(layer_data):\n",
    "             layer_has_nan = np.isnan(layer_data.data).any()\n",
    "             layer_has_inf = np.isinf(layer_data.data).any()\n",
    "             print(f\"NaN in {name}.layers['counts'].data: {layer_has_nan} (count: {np.isnan(layer_data.data).sum()})\")\n",
    "             print(f\"Inf in {name}.layers['counts'].data: {layer_has_inf} (count: {np.isinf(layer_data.data).sum()})\")\n",
    "        elif isinstance(layer_data, np.ndarray):\n",
    "             layer_has_nan = np.isnan(layer_data).any()\n",
    "             layer_has_inf = np.isinf(layer_data).any()\n",
    "             print(f\"NaN in {name}.layers['counts']: {layer_has_nan} (count: {np.isnan(layer_data).sum()})\")\n",
    "             print(f\"Inf in {name}.layers['counts']: {layer_has_inf} (count: {np.isinf(layer_data).sum()})\")\n",
    "\n",
    "\n",
    "check_nan_inf(st_data, \"st_data\")\n",
    "check_nan_inf(sc_data, \"sc_data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e35f147-d6bc-4d19-9e6d-5672827d1df1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c4df7249-0101-45bd-9e5a-dcb32643ce4c",
   "metadata": {},
   "source": [
    "# SpaGE spatial gene expression prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "92dc693f-0460-4110-ba8d-e994f39f4a73",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n",
      "/home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/tissue/main.py:377: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  predicted_expression_target[conf_genes] = predicted_expression_conf[conf_genes].copy()\n"
     ]
    }
   ],
   "source": [
    "# SpaGE spatial gene expression prediction\n",
    "\n",
    "tissue.main.predict_gene_expression (st_data, sc_data, [target_gene],method=\"spage\", n_folds=10, n_pv=10)\n",
    "# /home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
    "#   return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)\n",
    "# /home/jaecheon/anaconda3/envs/scenvi/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:4266: FutureWarning: The behavior of DataFrame.var with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
    "#   return var(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1238e0c-872e-4af8-b29a-c1cb0117d260",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SpaGE prediction vs actual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "19e14700-8bc2-423f-aba3-1e01b7367566",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "'c' argument has 62941 elements, which is inconsistent with 'x' and 'y' with size 61246.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/matplotlib/axes/_axes.py:4755\u001b[0m, in \u001b[0;36mAxes._parse_scatter_color_args\u001b[0;34m(c, edgecolors, kwargs, xsize, get_next_color_func)\u001b[0m\n\u001b[1;32m   4754\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:  \u001b[38;5;66;03m# Is 'c' acceptable as PathCollection facecolors?\u001b[39;00m\n\u001b[0;32m-> 4755\u001b[0m     colors \u001b[38;5;241m=\u001b[39m \u001b[43mmcolors\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_rgba_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43mc\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4756\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mTypeError\u001b[39;00m, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/matplotlib/colors.py:515\u001b[0m, in \u001b[0;36mto_rgba_array\u001b[0;34m(c, alpha)\u001b[0m\n\u001b[1;32m    514\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 515\u001b[0m     rgba \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray([to_rgba(cc) \u001b[38;5;28;01mfor\u001b[39;00m cc \u001b[38;5;129;01min\u001b[39;00m c])\n\u001b[1;32m    517\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m alpha \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/matplotlib/colors.py:515\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    514\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 515\u001b[0m     rgba \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray([\u001b[43mto_rgba\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcc\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m cc \u001b[38;5;129;01min\u001b[39;00m c])\n\u001b[1;32m    517\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m alpha \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/matplotlib/colors.py:317\u001b[0m, in \u001b[0;36mto_rgba\u001b[0;34m(c, alpha)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m rgba \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:  \u001b[38;5;66;03m# Suppress exception chaining of cache lookup failure.\u001b[39;00m\n\u001b[0;32m--> 317\u001b[0m     rgba \u001b[38;5;241m=\u001b[39m \u001b[43m_to_rgba_no_colorcycle\u001b[49m\u001b[43m(\u001b[49m\u001b[43mc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43malpha\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    318\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/matplotlib/colors.py:403\u001b[0m, in \u001b[0;36m_to_rgba_no_colorcycle\u001b[0;34m(c, alpha)\u001b[0m\n\u001b[1;32m    402\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(c) \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;241m3\u001b[39m, \u001b[38;5;241m4\u001b[39m]:\n\u001b[0;32m--> 403\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRGBA sequence should have length 3 or 4\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    404\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mall\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, Real) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m c):\n\u001b[1;32m    405\u001b[0m     \u001b[38;5;66;03m# Checks that don't work: `map(float, ...)`, `np.array(..., float)` and\u001b[39;00m\n\u001b[1;32m    406\u001b[0m     \u001b[38;5;66;03m# `np.array(...).astype(float)` would all convert \"0.5\" to 0.5.\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: RGBA sequence should have length 3 or 4",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[62], line 10\u001b[0m\n\u001b[1;32m      8\u001b[0m cmap \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mlog1p(cmap)\n\u001b[1;32m      9\u001b[0m cmap[cmap \u001b[38;5;241m>\u001b[39m np\u001b[38;5;241m.\u001b[39mpercentile(cmap,\u001b[38;5;241m95\u001b[39m)] \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mpercentile(cmap,\u001b[38;5;241m95\u001b[39m)\n\u001b[0;32m---> 10\u001b[0m im \u001b[38;5;241m=\u001b[39m \u001b[43max1\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mscatter\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobsm\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mspatial\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43madata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobsm\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mspatial\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43ms\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mc\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcmap\u001b[49m\u001b[43m,\u001b[49m\u001b[43mrasterized\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m     11\u001b[0m ax1\u001b[38;5;241m.\u001b[39mset_title(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mActual\u001b[39m\u001b[38;5;124m'\u001b[39m, fontsize \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m12\u001b[39m)\n\u001b[1;32m     13\u001b[0m cbar \u001b[38;5;241m=\u001b[39m fig\u001b[38;5;241m.\u001b[39mcolorbar(im)\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/matplotlib/_api/deprecation.py:453\u001b[0m, in \u001b[0;36mmake_keyword_only.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    447\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m name_idx:\n\u001b[1;32m    448\u001b[0m     warn_deprecated(\n\u001b[1;32m    449\u001b[0m         since, message\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPassing the \u001b[39m\u001b[38;5;132;01m%(name)s\u001b[39;00m\u001b[38;5;124m \u001b[39m\u001b[38;5;132;01m%(obj_type)s\u001b[39;00m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    450\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpositionally is deprecated since Matplotlib \u001b[39m\u001b[38;5;132;01m%(since)s\u001b[39;00m\u001b[38;5;124m; the \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    451\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparameter will become keyword-only in \u001b[39m\u001b[38;5;132;01m%(removal)s\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    452\u001b[0m         name\u001b[38;5;241m=\u001b[39mname, obj_type\u001b[38;5;241m=\u001b[39m\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparameter of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfunc\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m()\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 453\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/matplotlib/__init__.py:1521\u001b[0m, in \u001b[0;36m_preprocess_data.<locals>.inner\u001b[0;34m(ax, data, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1518\u001b[0m \u001b[38;5;129m@functools\u001b[39m\u001b[38;5;241m.\u001b[39mwraps(func)\n\u001b[1;32m   1519\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21minner\u001b[39m(ax, \u001b[38;5;241m*\u001b[39margs, data\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m   1520\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m data \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1521\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1522\u001b[0m \u001b[43m            \u001b[49m\u001b[43max\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1523\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mmap\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mcbook\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msanitize_sequence\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1524\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43m{\u001b[49m\u001b[43mk\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mcbook\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msanitize_sequence\u001b[49m\u001b[43m(\u001b[49m\u001b[43mv\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mitems\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1526\u001b[0m     bound \u001b[38;5;241m=\u001b[39m new_sig\u001b[38;5;241m.\u001b[39mbind(ax, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m   1527\u001b[0m     auto_label \u001b[38;5;241m=\u001b[39m (bound\u001b[38;5;241m.\u001b[39marguments\u001b[38;5;241m.\u001b[39mget(label_namer)\n\u001b[1;32m   1528\u001b[0m                   \u001b[38;5;129;01mor\u001b[39;00m bound\u001b[38;5;241m.\u001b[39mkwargs\u001b[38;5;241m.\u001b[39mget(label_namer))\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/matplotlib/axes/_axes.py:4948\u001b[0m, in \u001b[0;36mAxes.scatter\u001b[0;34m(self, x, y, s, c, marker, cmap, norm, vmin, vmax, alpha, linewidths, edgecolors, colorizer, plotnonfinite, **kwargs)\u001b[0m\n\u001b[1;32m   4945\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m edgecolors \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   4946\u001b[0m     orig_edgecolor \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124medgecolor\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m   4947\u001b[0m c, colors, edgecolors \u001b[38;5;241m=\u001b[39m \\\n\u001b[0;32m-> 4948\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_parse_scatter_color_args\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   4949\u001b[0m \u001b[43m        \u001b[49m\u001b[43mc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43medgecolors\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4950\u001b[0m \u001b[43m        \u001b[49m\u001b[43mget_next_color_func\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_patches_for_fill\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_next_color\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4952\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m plotnonfinite \u001b[38;5;129;01mand\u001b[39;00m colors \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   4953\u001b[0m     c \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mma\u001b[38;5;241m.\u001b[39mmasked_invalid(c)\n",
      "File \u001b[0;32m~/anaconda3/envs/scenvi/lib/python3.10/site-packages/matplotlib/axes/_axes.py:4761\u001b[0m, in \u001b[0;36mAxes._parse_scatter_color_args\u001b[0;34m(c, edgecolors, kwargs, xsize, get_next_color_func)\u001b[0m\n\u001b[1;32m   4759\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   4760\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m valid_shape:\n\u001b[0;32m-> 4761\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m invalid_shape_exception(c\u001b[38;5;241m.\u001b[39msize, xsize) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   4762\u001b[0m     \u001b[38;5;66;03m# Both the mapping *and* the RGBA conversion failed: pretty\u001b[39;00m\n\u001b[1;32m   4763\u001b[0m     \u001b[38;5;66;03m# severe failure => one may appreciate a verbose feedback.\u001b[39;00m\n\u001b[1;32m   4764\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   4765\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mc\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m argument must be a color, a sequence of colors, \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   4766\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mor a sequence of numbers, not \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mc\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01merr\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: 'c' argument has 62941 elements, which is inconsistent with 'x' and 'y' with size 61246."
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAg8AAAGiCAYAAABgTyUPAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAAG3lJREFUeJzt3W1s1eX9+PFPW+wpRlvZGOVmdUw35z04kK4648+lWxMNGw8WmRhgxJs5mXE0m4IonVMpc0rIBEdkOvdgG0yjyzIITjuJUVmIQBOdqHGosGWtMkfLqrbSfn8P9vf8/pWiXIy2x53XKzkPenl9z/kcw2ne/Z67kizLsgAAOESlwz0AAPDRIh4AgCTiAQBIIh4AgCTiAQBIIh4AgCTiAQBIIh4AgCTiAQBIIh4AgCTiAYrAE088EdOnT4/x48dHSUlJ/Pa3v/3QYzZt2hSf//znI5fLxWc+85m4//77B31O4KNBPEAR6OrqikmTJsWqVasOaf8rr7wSF110UVxwwQXR2toa3/3ud+Pyyy+PRx55ZJAnBT4KSnwxFhSXkpKSePjhh2PGjBkH3XP99dfH+vXr47nnnsuvfeMb34i9e/fGxo0bh2BKoJCNGO4BgMKzefPmqK+v77fW0NAQ3/3udz/wuO7u7uju7s7/3NfXF2+++WZ8/OMfj5KSksEYFfgAWZbFvn37Yvz48VFaeuSebBAPwAHa2tqiurq631p1dXV0dnbG22+/HSNHjhzwuObm5rj55puHYkQgwe7du+OTn/zkEbs+8QAcMYsWLYrGxsb8zx0dHXH88cfH7t27o7Kychgng+LU2dkZNTU1ceyxxx7R6xUPwAHGjh0b7e3t/dba29ujsrLyoGcdIiJyuVzkcrkD1isrK8UDDKMj/bShd1sAB6irq4uWlpZ+a48++mjU1dUN00RAIREPUAT+9a9/RWtra7S2tkbEv9+K2draGrt27YqIfz/dMGfOnPz+q666Knbu3BnXXXddvPDCC3H33XfHb37zm1iwYMFwjA8UGPEAReCZZ56Js846K84666yIiGhsbIyzzjorlixZEhERf//73/MhERHx6U9/OtavXx+PPvpoTJo0Ke6888742c9+Fg0NDcMyP1BYfM4DMGg6OzujqqoqOjo6vOYBhsFgPQadeQAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHKCKrVq2KiRMnRkVFRdTW1saWLVs+cP+KFSvic5/7XIwcOTJqampiwYIF8c477wzRtEChEg9QJNatWxeNjY3R1NQU27Zti0mTJkVDQ0O8/vrrA+7/1a9+FQsXLoympqbYsWNH3HvvvbFu3bq44YYbhnhyoNCIBygSy5cvjyuuuCLmzZsXp556aqxevTqOPvrouO+++wbc//TTT8e5554bs2bNiokTJ8ZXvvKVuOSSSz70bAXw3088QBHo6emJrVu3Rn19fX6ttLQ06uvrY/PmzQMec84558TWrVvzsbBz587YsGFDXHjhhQe9ne7u7ujs7Ox3Af77jBjuAYDBt2fPnujt7Y3q6up+69XV1fHCCy8MeMysWbNiz5498cUvfjGyLIv9+/fHVVdd9YFPWzQ3N8fNN998RGcHCo8zD8CANm3aFEuXLo277747tm3bFg899FCsX78+brnlloMes2jRoujo6Mhfdu/ePYQTA0PFmQcoAqNHj46ysrJob2/vt97e3h5jx44d8JibbropZs+eHZdffnlERJxxxhnR1dUVV155ZSxevDhKSw/82yOXy0UulzvydwAoKM48QBEoLy+PKVOmREtLS36tr68vWlpaoq6ubsBj3nrrrQMCoaysLCIisiwbvGGBgufMAxSJxsbGmDt3bkydOjWmTZsWK1asiK6urpg3b15ERMyZMycmTJgQzc3NERExffr0WL58eZx11llRW1sbL7/8ctx0000xffr0fEQAxUk8QJGYOXNmvPHGG7FkyZJoa2uLyZMnx8aNG/Mvoty1a1e/Mw033nhjlJSUxI033hh/+9vf4hOf+ERMnz49brvttuG6C0CBKMmcfwQGSWdnZ1RVVUVHR0dUVlYO9zhQdAbrMeg1DwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEAACQRDwBAEvEARWTVqlUxceLEqKioiNra2tiyZcsH7t+7d2/Mnz8/xo0bF7lcLk466aTYsGHDEE0LFKoRwz0AMDTWrVsXjY2NsXr16qitrY0VK1ZEQ0NDvPjiizFmzJgD9vf09MSXv/zlGDNmTDz44IMxYcKEeO211+K4444b+uGBglKSZVk23EMAg6+2tjbOPvvsWLlyZURE9PX1RU1NTVxzzTWxcOHCA/avXr06fvzjH8cLL7wQRx111GHdZmdnZ1RVVUVHR0dUVlb+R/MD6QbrMehpCygCPT09sXXr1qivr8+vlZaWRn19fWzevHnAY373u99FXV1dzJ8/P6qrq+P000+PpUuXRm9v70Fvp7u7Ozo7O/tdgP8+4gGKwJ49e6K3tzeqq6v7rVdXV0dbW9uAx+zcuTMefPDB6O3tjQ0bNsRNN90Ud955Z9x6660HvZ3m5uaoqqrKX2pqao7o/QAKg3gABtTX1xdjxoyJe+65J6ZMmRIzZ86MxYsXx+rVqw96zKJFi6KjoyN/2b179xBODAwVL5iEIjB69OgoKyuL9vb2fuvt7e0xduzYAY8ZN25cHHXUUVFWVpZfO+WUU6KtrS16enqivLz8gGNyuVzkcrkjOzxQcJx5gCJQXl4eU6ZMiZaWlvxaX19ftLS0RF1d3YDHnHvuufHyyy9HX19ffu2ll16KcePGDRgOQPEQD1AkGhsbY82aNfGLX/widuzYEd/+9rejq6sr5s2bFxERc+bMiUWLFuX3f/vb344333wzrr322njppZdi/fr1sXTp0pg/f/5w3QWgQHjaAorEzJkz44033oglS5ZEW1tbTJ48OTZu3Jh/EeWuXbuitPT//p6oqamJRx55JBYsWBBnnnlmTJgwIa699tq4/vrrh+suAAXC5zwAg8bnPMDw8jkPAEBBEA8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ8AQBLxAAAkEQ9QRFatWhUTJ06MioqKqK2tjS1bthzScWvXro2SkpKYMWPG4A4IfCSIBygS69ati8bGxmhqaopt27bFpEmToqGhIV5//fUPPO7VV1+N733ve3HeeecN0aRAoRMPUCSWL18eV1xxRcybNy9OPfXUWL16dRx99NFx3333HfSY3t7euPTSS+Pmm2+OE0444UNvo7u7Ozo7O/tdgP8+4gGKQE9PT2zdujXq6+vza6WlpVFfXx+bN28+6HE//OEPY8yYMXHZZZcd0u00NzdHVVVV/lJTU/Mfzw4UHvEARWDPnj3R29sb1dXV/darq6ujra1twGOefPLJuPfee2PNmjWHfDuLFi2Kjo6O/GX37t3/0dxAYRox3AMAhWffvn0xe/bsWLNmTYwePfqQj8vlcpHL5QZxMqAQiAcoAqNHj46ysrJob2/vt97e3h5jx449YP9f/vKXePXVV2P69On5tb6+voiIGDFiRLz44otx4oknDu7QQMHytAUUgfLy8pgyZUq0tLTk1/r6+qKlpSXq6uoO2H/yySfHs88+G62trfnLV7/61bjggguitbXVaxmgyDnzAEWisbEx5s6dG1OnTo1p06bFihUroqurK+bNmxcREXPmzIkJEyZEc3NzVFRUxOmnn97v+OOOOy4i4oB1oPiIBygSM2fOjDfeeCOWLFkSbW1tMXny5Ni4cWP+RZS7du2K0lInI4EPV5JlWTbcQwD/nTo7O6Oqqio6OjqisrJyuMeBojNYj0F/ZgAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDAJBEPAAAScQDFJFVq1bFxIkTo6KiImpra2PLli0H3btmzZo477zzYtSoUTFq1Kior6//wP1A8RAPUCTWrVsXjY2N0dTUFNu2bYtJkyZFQ0NDvP766wPu37RpU1xyySXx+OOPx+bNm6Ompia+8pWvxN/+9rchnhwoNCVZlmXDPQQw+Gpra+Pss8+OlStXRkREX19f1NTUxDXXXBMLFy780ON7e3tj1KhRsXLlypgzZ86Ae7q7u6O7uzv/c2dnZ9TU1ERHR0dUVlYemTsCHLLOzs6oqqo64o9BZx6gCPT09MTWrVujvr4+v1ZaWhr19fWxefPmQ7qOt956K95999342Mc+dtA9zc3NUVVVlb/U1NT8x7MDhUc8QBHYs2dP9Pb2RnV1db/16urqaGtrO6TruP7662P8+PH9AuT9Fi1aFB0dHfnL7t27/6O5gcI0YrgHAArfsmXLYu3atbFp06aoqKg46L5cLhe5XG4IJwOGg3iAIjB69OgoKyuL9vb2fuvt7e0xduzYDzz2jjvuiGXLlsVjjz0WZ5555mCOCXxEeNoCikB5eXlMmTIlWlpa8mt9fX3R0tISdXV1Bz3u9ttvj1tuuSU2btwYU6dOHYpRgY8AZx6gSDQ2NsbcuXNj6tSpMW3atFixYkV0dXXFvHnzIiJizpw5MWHChGhubo6IiB/96EexZMmS+NWvfhUTJ07MvzbimGOOiWOOOWbY7gcw/MQDFImZM2fGG2+8EUuWLIm2traYPHlybNy4Mf8iyl27dkVp6f+djPzpT38aPT098fWvf73f9TQ1NcUPfvCDoRwdKDA+5wEYNIP1HnPg0PicBwCgIIgHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHACCJeAAAkogHKCKrVq2KiRMnRkVFRdTW1saWLVs+cP8DDzwQJ598clRUVMQZZ5wRGzZsGKJJgUImHqBIrFu3LhobG6OpqSm2bdsWkyZNioaGhnj99dcH3P/000/HJZdcEpdddlls3749ZsyYETNmzIjnnntuiCcHCk1JlmXZcA8BDL7a2to4++yzY+XKlRER0dfXFzU1NXHNNdfEwoULD9g/c+bM6Orqit///vf5tS984QsxefLkWL169YC30d3dHd3d3fmfOzo64vjjj4/du3dHZWXlEb5HwIfp7OyMmpqa2Lt3b1RVVR2x6x1xxK4JKFg9PT2xdevWWLRoUX6ttLQ06uvrY/PmzQMes3nz5mhsbOy31tDQEL/97W8PejvNzc1x8803H7BeU1NzeIMDR8Q//vEP8QCk2bNnT/T29kZ1dXW/9erq6njhhRcGPKatrW3A/W1tbQe9nUWLFvULjr1798anPvWp2LVr1xH9xTWY3vtL7aN0tsTMQ+OjOPN7Z/8+9rGPHdHrFQ/AEZPL5SKXyx2wXlVV9ZH5ZfueyspKMw8BMw+N0tIj+xJHL5iEIjB69OgoKyuL9vb2fuvt7e0xduzYAY8ZO3Zs0n6geIgHKALl5eUxZcqUaGlpya/19fVFS0tL1NXVDXhMXV1dv/0REY8++uhB9wPFw9MWUCQaGxtj7ty5MXXq1Jg2bVqsWLEiurq6Yt68eRERMWfOnJgwYUI0NzdHRMS1114b559/ftx5551x0UUXxdq1a+OZZ56Je+6555BvM5fLRVNT04BPZRQqMw8NMw+NwZrZWzWhiKxcuTJ+/OMfR1tbW0yePDl+8pOfRG1tbURE/M///E9MnDgx7r///vz+Bx54IG688cZ49dVX47Of/WzcfvvtceGFFw7T9EChEA8AQBKveQAAkogHACCJeAAAkogHACCJeAAO20f1K75T5l6zZk2cd955MWrUqBg1alTU19d/6P0cDKn/r9+zdu3aKCkpiRkzZgzugANInXnv3r0xf/78GDduXORyuTjppJOG/N9I6swrVqyIz33uczFy5MioqamJBQsWxDvvvDNE00Y88cQTMX369Bg/fnyUlJR84HfPvGfTpk3x+c9/PnK5XHzmM5/p9w6rQ5YBHIa1a9dm5eXl2X333Zf9+c9/zq644orsuOOOy9rb2wfc/9RTT2VlZWXZ7bffnj3//PPZjTfemB111FHZs88+W9Bzz5o1K1u1alW2ffv2bMeOHdk3v/nNrKqqKvvrX/9asDO/55VXXskmTJiQnXfeednXvva1oRn2/0mdubu7O5s6dWp24YUXZk8++WT2yiuvZJs2bcpaW1sLduZf/vKXWS6Xy375y19mr7zySvbII49k48aNyxYsWDBkM2/YsCFbvHhx9tBDD2URkT388MMfuH/nzp3Z0UcfnTU2NmbPP/98dtddd2VlZWXZxo0bk25XPACHZdq0adn8+fPzP/f29mbjx4/PmpubB9x/8cUXZxdddFG/tdra2uxb3/rWoM75fqlzv9/+/fuzY489NvvFL34xWCMe4HBm3r9/f3bOOedkP/vZz7K5c+cOeTykzvzTn/40O+GEE7Kenp6hGvEAqTPPnz8/+9KXvtRvrbGxMTv33HMHdc6DOZR4uO6667LTTjut39rMmTOzhoaGpNvytAWQ7L2v+K6vr8+vHcpXfP//+yP+/RXfB9s/GA5n7vd766234t133z3i31J4MIc78w9/+MMYM2ZMXHbZZUMxZj+HM/Pvfve7qKuri/nz50d1dXWcfvrpsXTp0ujt7S3Ymc8555zYunVr/qmNnTt3xoYNGwr6g9SO1OPQx1MDyYbqK76PtMOZ+/2uv/76GD9+/AG/gAfL4cz85JNPxr333hutra1DMOGBDmfmnTt3xh//+Me49NJLY8OGDfHyyy/H1VdfHe+++240NTUV5MyzZs2KPXv2xBe/+MXIsiz2798fV111Vdxwww2DPu/hOtjjsLOzM95+++0YOXLkIV2PMw8Ah2jZsmWxdu3aePjhh6OiomK4xxnQvn37Yvbs2bFmzZoYPXr0cI9zyPr6+mLMmDFxzz33xJQpU2LmzJmxePHiWL169XCPdlCbNm2KpUuXxt133x3btm2Lhx56KNavXx+33HLLcI826Jx5AJJ9VL/i+3Dmfs8dd9wRy5Yti8ceeyzOPPPMwRyzn9SZ//KXv8Srr74a06dPz6/19fVFRMSIESPixRdfjBNPPLGgZo6IGDduXBx11FFRVlaWXzvllFOira0tenp6ory8vOBmvummm2L27Nlx+eWXR0TEGWecEV1dXXHllVfG4sWLo7S08P4+P9jjsLKy8pDPOkQ48wAcho/qV3wfztwREbfffnvccsstsXHjxpg6depQjJqXOvPJJ58czz77bLS2tuYvX/3qV+OCCy6I1tbWqKmpKbiZIyLOPffcePnll/OhExHx0ksvxbhx4wY9HA535rfeeuuAQHgvfrIC/dqoI/Y4THstJ8C/rV27Nsvlctn999+fPf/889mVV16ZHXfccVlbW1uWZVk2e/bsbOHChfn9Tz31VDZixIjsjjvuyHbs2JE1NTUN21s1U+ZetmxZVl5enj344IPZ3//+9/xl3759BTvz+w3Huy1SZ961a1d27LHHZt/5zneyF198Mfv973+fjRkzJrv11lsLduampqbs2GOPzX79619nO3fuzP7whz9kJ554YnbxxRcP2cz79u3Ltm/fnm3fvj2LiGz58uXZ9u3bs9deey3LsixbuHBhNnv27Pz+996q+f3vfz/bsWNHtmrVKm/VBIbWXXfdlR1//PFZeXl5Nm3atOxPf/pT/r+df/752dy5c/vt/81vfpOddNJJWXl5eXbaaadl69evH+KJ/y1l7k996lNZRBxwaWpqKtiZ32844iHL0md++umns9ra2iyXy2UnnHBCdtttt2X79+8v2Jnffffd7Ac/+EF24oknZhUVFVlNTU129dVXZ//85z+HbN7HH398wH+f7805d+7c7Pzzzz/gmMmTJ2fl5eXZCSeckP385z9Pvl1fyQ0AJPGaBwAgiXgAAJKIBwAgiXgAAJKIBwAgiXgAAJKIBwAgiXgAAJKIBwAgiXgAAJKIBwAgyf8CN822QaA11u8AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Visualizing predicted and actual expression side by side\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "\n",
    "ax1.axis('off')\n",
    "cmap = target_expn\n",
    "cmap[cmap<0] = 0\n",
    "cmap = np.log1p(cmap)\n",
    "cmap[cmap > np.percentile(cmap,95)] = np.percentile(cmap,95)\n",
    "im = ax1.scatter(adata.obsm['spatial'][:,0],adata.obsm['spatial'][:,1],s=1,c=cmap,rasterized=True)\n",
    "ax1.set_title('Actual', fontsize = 12)\n",
    "\n",
    "cbar = fig.colorbar(im)\n",
    "cbar.ax.get_yaxis().labelpad = 15\n",
    "cbar.ax.set_ylabel('Log Expression', rotation=270)\n",
    "\n",
    "ax2.axis('off')\n",
    "cmap = adata.obsm['spage_predicted_expression'][target_gene].values\n",
    "cmap[cmap<0] = 0\n",
    "cmap = np.log1p(cmap)\n",
    "cmap[cmap > np.percentile(cmap,95)] = np.percentile(cmap,95)\n",
    "im = ax2.scatter(adata.obsm['spatial'][:,0],adata.obsm['spatial'][:,1],s=1,c=cmap,rasterized=True)\n",
    "ax2.set_title('Predicted', fontsize = 12)\n",
    "\n",
    "cbar = fig.colorbar(im)\n",
    "cbar.ax.get_yaxis().labelpad = 15\n",
    "cbar.ax.set_ylabel('Log Expression', rotation=270)\n",
    "\n",
    "plt.suptitle(\"SpaGE Prediction\", fontsize=16)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b5b5d40b-4527-4e7c-94cd-a314b9561dab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7da081e1-b94a-44d9-ace3-f7b7ad43afea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
